<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.361">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Laura’s Blog - Import Libraries</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Laura’s Blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com" rel="" target=""><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Import Libraries</h1>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<div class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> tensorflow <span class="im">as</span> tf</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Make sure the GPU is enabled </span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> tf.config.list_physical_devices(<span class="st">'GPU'</span>), <span class="st">'Start the colab kernel with GPU: Runtime -&gt; Change runtime type -&gt; GPU'</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Num GPUs Available: "</span>, <span class="bu">len</span>(tf.config.list_physical_devices(<span class="st">'GPU'</span>)))</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> tensorflow_hub <span class="im">as</span> hub <span class="co"># for interacting with saved models and tensorflow hub</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> joblib</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> gzip <span class="co"># for manipulating compressed files</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> kipoiseq <span class="co"># for manipulating fasta files</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> kipoiseq <span class="im">import</span> Interval <span class="co"># same as above, really</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyfaidx <span class="co"># to index our reference genome file</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd <span class="co"># for manipulating dataframes</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np <span class="co"># for numerical computations</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt <span class="co"># for plotting</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib <span class="im">as</span> mpl <span class="co"># for plotting</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns <span class="co"># for plotting</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pickle <span class="co"># for saving large objects</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os, sys <span class="co"># functions for interacting with the operating system</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collections <span class="im">import</span> Counter</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>matplotlib inline</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>config InlineBackend.figure_format <span class="op">=</span> <span class="st">'retina'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>2023-07-17 19:17:14.481091: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Num GPUs Available:  4</code></pre>
</div>
</div>
<section id="define-paths" class="level2">
<h2 class="anchored" data-anchor-id="define-paths">Define Paths</h2>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>transform_path <span class="op">=</span> <span class="st">'gs://dm-enformer/models/enformer.finetuned.SAD.robustscaler-PCA500-robustscaler.transform.pkl'</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>model_path <span class="op">=</span> <span class="st">'https://tfhub.dev/deepmind/enformer/1'</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>fasta_file <span class="op">=</span> <span class="st">'/grand/TFXcan/imlab/users/lvairus/hackenf/data/genome.fa'</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>targets_txt <span class="op">=</span> <span class="st">'https://raw.githubusercontent.com/calico/basenji/master/manuscripts/cross2020/targets_human.txt'</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>df_targets <span class="op">=</span> pd.read_csv(targets_txt, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="define-functions" class="level2">
<h2 class="anchored" data-anchor-id="define-functions">Define Functions</h2>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># @title `Enformer`, `EnformerScoreVariantsNormalized`, `EnformerScoreVariantsPCANormalized`,</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>SEQUENCE_LENGTH <span class="op">=</span> <span class="dv">393216</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> Enformer:</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, tfhub_url):</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>._model <span class="op">=</span> hub.load(tfhub_url).model</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> predict_on_batch(<span class="va">self</span>, inputs):</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>    predictions <span class="op">=</span> <span class="va">self</span>._model.predict_on_batch(inputs)</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> {k: v.numpy() <span class="cf">for</span> k, v <span class="kw">in</span> predictions.items()}</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">@tf.function</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> contribution_input_grad(<span class="va">self</span>, input_sequence,</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>                              target_mask, output_head<span class="op">=</span><span class="st">'human'</span>):</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>    input_sequence <span class="op">=</span> input_sequence[tf.newaxis]</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>    target_mask_mass <span class="op">=</span> tf.reduce_sum(target_mask)</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> tf.GradientTape() <span class="im">as</span> tape:</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>      tape.watch(input_sequence)</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>      prediction <span class="op">=</span> tf.reduce_sum(</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>          target_mask[tf.newaxis] <span class="op">*</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>          <span class="va">self</span>._model.predict_on_batch(input_sequence)[output_head]) <span class="op">/</span> target_mask_mass</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>    input_grad <span class="op">=</span> tape.gradient(prediction, input_sequence) <span class="op">*</span> input_sequence</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>    input_grad <span class="op">=</span> tf.squeeze(input_grad, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> tf.reduce_sum(input_grad, axis<span class="op">=-</span><span class="dv">1</span>)</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> EnformerScoreVariantsRaw:</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, tfhub_url, organism<span class="op">=</span><span class="st">'human'</span>):</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>._model <span class="op">=</span> Enformer(tfhub_url)</span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>._organism <span class="op">=</span> organism</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> predict_on_batch(<span class="va">self</span>, inputs):</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>    ref_prediction <span class="op">=</span> <span class="va">self</span>._model.predict_on_batch(inputs[<span class="st">'ref'</span>])[<span class="va">self</span>._organism]</span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a>    alt_prediction <span class="op">=</span> <span class="va">self</span>._model.predict_on_batch(inputs[<span class="st">'alt'</span>])[<span class="va">self</span>._organism]</span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> alt_prediction.mean(axis<span class="op">=</span><span class="dv">1</span>) <span class="op">-</span> ref_prediction.mean(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> EnformerScoreVariantsNormalized:</span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, tfhub_url, transform_pkl_path,</span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a>               organism<span class="op">=</span><span class="st">'human'</span>):</span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a>    <span class="cf">assert</span> organism <span class="op">==</span> <span class="st">'human'</span>, <span class="st">'Transforms only compatible with organism=human'</span></span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>._model <span class="op">=</span> EnformerScoreVariantsRaw(tfhub_url, organism)</span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> tf.io.gfile.GFile(transform_pkl_path, <span class="st">'rb'</span>) <span class="im">as</span> f:</span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a>      transform_pipeline <span class="op">=</span> joblib.load(f)</span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>._transform <span class="op">=</span> transform_pipeline.steps[<span class="dv">0</span>][<span class="dv">1</span>]  <span class="co"># StandardScaler.</span></span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> predict_on_batch(<span class="va">self</span>, inputs):</span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a>    scores <span class="op">=</span> <span class="va">self</span>._model.predict_on_batch(inputs)</span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="va">self</span>._transform.transform(scores)</span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> EnformerScoreVariantsPCANormalized:</span>
<span id="cb5-59"><a href="#cb5-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-60"><a href="#cb5-60" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, tfhub_url, transform_pkl_path,</span>
<span id="cb5-61"><a href="#cb5-61" aria-hidden="true" tabindex="-1"></a>               organism<span class="op">=</span><span class="st">'human'</span>, num_top_features<span class="op">=</span><span class="dv">500</span>):</span>
<span id="cb5-62"><a href="#cb5-62" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>._model <span class="op">=</span> EnformerScoreVariantsRaw(tfhub_url, organism)</span>
<span id="cb5-63"><a href="#cb5-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> tf.io.gfile.GFile(transform_pkl_path, <span class="st">'rb'</span>) <span class="im">as</span> f:</span>
<span id="cb5-64"><a href="#cb5-64" aria-hidden="true" tabindex="-1"></a>      <span class="va">self</span>._transform <span class="op">=</span> joblib.load(f)</span>
<span id="cb5-65"><a href="#cb5-65" aria-hidden="true" tabindex="-1"></a>    <span class="va">self</span>._num_top_features <span class="op">=</span> num_top_features</span>
<span id="cb5-66"><a href="#cb5-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-67"><a href="#cb5-67" aria-hidden="true" tabindex="-1"></a>  <span class="kw">def</span> predict_on_batch(<span class="va">self</span>, inputs):</span>
<span id="cb5-68"><a href="#cb5-68" aria-hidden="true" tabindex="-1"></a>    scores <span class="op">=</span> <span class="va">self</span>._model.predict_on_batch(inputs)</span>
<span id="cb5-69"><a href="#cb5-69" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="va">self</span>._transform.transform(scores)[:, :<span class="va">self</span>._num_top_features]</span>
<span id="cb5-70"><a href="#cb5-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-71"><a href="#cb5-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-72"><a href="#cb5-72" aria-hidden="true" tabindex="-1"></a><span class="co"># </span><span class="al">TODO</span><span class="co">(avsec): Add feature description: Either PCX, or full names.</span></span>
<span id="cb5-73"><a href="#cb5-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-74"><a href="#cb5-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-75"><a href="#cb5-75" aria-hidden="true" tabindex="-1"></a><span class="co"># @title `variant_centered_sequences`</span></span>
<span id="cb5-76"><a href="#cb5-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-77"><a href="#cb5-77" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> FastaStringExtractor:</span>
<span id="cb5-78"><a href="#cb5-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-79"><a href="#cb5-79" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, fasta_file):</span>
<span id="cb5-80"><a href="#cb5-80" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.fasta <span class="op">=</span> pyfaidx.Fasta(fasta_file)</span>
<span id="cb5-81"><a href="#cb5-81" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>._chromosome_sizes <span class="op">=</span> {k: <span class="bu">len</span>(v) <span class="cf">for</span> k, v <span class="kw">in</span> <span class="va">self</span>.fasta.items()}</span>
<span id="cb5-82"><a href="#cb5-82" aria-hidden="true" tabindex="-1"></a>    <span class="co">#import pd.Interval as Interval</span></span>
<span id="cb5-83"><a href="#cb5-83" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> extract(<span class="va">self</span>, interval: Interval, <span class="op">**</span>kwargs) <span class="op">-&gt;</span> <span class="bu">str</span>:</span>
<span id="cb5-84"><a href="#cb5-84" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Truncate interval if it extends beyond the chromosome lengths.</span></span>
<span id="cb5-85"><a href="#cb5-85" aria-hidden="true" tabindex="-1"></a>        chromosome_length <span class="op">=</span> <span class="va">self</span>._chromosome_sizes[interval.chrom]</span>
<span id="cb5-86"><a href="#cb5-86" aria-hidden="true" tabindex="-1"></a>        trimmed_interval <span class="op">=</span> Interval(interval.chrom,</span>
<span id="cb5-87"><a href="#cb5-87" aria-hidden="true" tabindex="-1"></a>                                    <span class="bu">max</span>(interval.start, <span class="dv">0</span>),</span>
<span id="cb5-88"><a href="#cb5-88" aria-hidden="true" tabindex="-1"></a>                                    <span class="bu">min</span>(interval.end, chromosome_length),</span>
<span id="cb5-89"><a href="#cb5-89" aria-hidden="true" tabindex="-1"></a>                                    )</span>
<span id="cb5-90"><a href="#cb5-90" aria-hidden="true" tabindex="-1"></a>        <span class="co"># pyfaidx wants a 1-based interval</span></span>
<span id="cb5-91"><a href="#cb5-91" aria-hidden="true" tabindex="-1"></a>        sequence <span class="op">=</span> <span class="bu">str</span>(<span class="va">self</span>.fasta.get_seq(trimmed_interval.chrom,</span>
<span id="cb5-92"><a href="#cb5-92" aria-hidden="true" tabindex="-1"></a>                                          trimmed_interval.start <span class="op">+</span> <span class="dv">1</span>,</span>
<span id="cb5-93"><a href="#cb5-93" aria-hidden="true" tabindex="-1"></a>                                          trimmed_interval.stop).seq).upper()</span>
<span id="cb5-94"><a href="#cb5-94" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Fill truncated values with N's.</span></span>
<span id="cb5-95"><a href="#cb5-95" aria-hidden="true" tabindex="-1"></a>        pad_upstream <span class="op">=</span> <span class="st">'N'</span> <span class="op">*</span> <span class="bu">max</span>(<span class="op">-</span>interval.start, <span class="dv">0</span>)</span>
<span id="cb5-96"><a href="#cb5-96" aria-hidden="true" tabindex="-1"></a>        pad_downstream <span class="op">=</span> <span class="st">'N'</span> <span class="op">*</span> <span class="bu">max</span>(interval.end <span class="op">-</span> chromosome_length, <span class="dv">0</span>)</span>
<span id="cb5-97"><a href="#cb5-97" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> pad_upstream <span class="op">+</span> sequence <span class="op">+</span> pad_downstream</span>
<span id="cb5-98"><a href="#cb5-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-99"><a href="#cb5-99" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> close(<span class="va">self</span>):</span>
<span id="cb5-100"><a href="#cb5-100" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="va">self</span>.fasta.close()</span>
<span id="cb5-101"><a href="#cb5-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-102"><a href="#cb5-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-103"><a href="#cb5-103" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> one_hot_encode(sequence):</span>
<span id="cb5-104"><a href="#cb5-104" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> kipoiseq.transforms.functional.one_hot_dna(sequence).astype(np.float32)</span>
<span id="cb5-105"><a href="#cb5-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-106"><a href="#cb5-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-107"><a href="#cb5-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-108"><a href="#cb5-108" aria-hidden="true" tabindex="-1"></a><span class="co"># @title `plot_tracks`</span></span>
<span id="cb5-109"><a href="#cb5-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-110"><a href="#cb5-110" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_tracks(tracks, interval, height<span class="op">=</span><span class="fl">1.5</span>):</span>
<span id="cb5-111"><a href="#cb5-111" aria-hidden="true" tabindex="-1"></a>  fig, axes <span class="op">=</span> plt.subplots(<span class="bu">len</span>(tracks), <span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">20</span>, height <span class="op">*</span> <span class="bu">len</span>(tracks)), sharex<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb5-112"><a href="#cb5-112" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> ax, (title, y) <span class="kw">in</span> <span class="bu">zip</span>(axes, tracks.items()):</span>
<span id="cb5-113"><a href="#cb5-113" aria-hidden="true" tabindex="-1"></a>    ax.fill_between(np.linspace(interval.start, interval.end, num<span class="op">=</span><span class="bu">len</span>(y)), y)</span>
<span id="cb5-114"><a href="#cb5-114" aria-hidden="true" tabindex="-1"></a>    ax.set_title(title)</span>
<span id="cb5-115"><a href="#cb5-115" aria-hidden="true" tabindex="-1"></a>    sns.despine(top<span class="op">=</span><span class="va">True</span>, right<span class="op">=</span><span class="va">True</span>, bottom<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb5-116"><a href="#cb5-116" aria-hidden="true" tabindex="-1"></a>  ax.set_xlabel(<span class="bu">str</span>(interval))</span>
<span id="cb5-117"><a href="#cb5-117" aria-hidden="true" tabindex="-1"></a>  plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> Bio</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> Bio.Seq <span class="im">import</span> Seq</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> create_rev_complement(dna_string):</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>(<span class="bu">str</span>(Seq(dna_string).reverse_complement()))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> prepare_for_quantify_prediction_per_TSS(predictions, gene, tss_df):</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="co">'''</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co">  Parameters:</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co">          predicitions (A numpy array): All predictions from the track</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co">          gene (a gene name, character): a gene</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co">          tss_df: a list of dataframe of genes and their transcription start sites</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co">  Returns:</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="co">          A dictionary of cage experiment predictions and a list of transcription start sites</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="co">  '''</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>  output <span class="op">=</span> <span class="bu">dict</span>()</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> tdf <span class="kw">in</span> tss_df:</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> gene <span class="kw">not</span> <span class="kw">in</span> tdf.genes.values:</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>      <span class="cf">continue</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>    gene_tss_list <span class="op">=</span> tdf[tdf.genes <span class="op">==</span> gene].txStart_Sites.<span class="bu">apply</span>(<span class="bu">str</span>).values</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>    gene_tss_list <span class="op">=</span> [t.split(<span class="st">', '</span>) <span class="cf">for</span> t <span class="kw">in</span> gene_tss_list]</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>    gene_tss_list <span class="op">=</span> [<span class="bu">int</span>(item) <span class="cf">for</span> nestedlist <span class="kw">in</span> gene_tss_list <span class="cf">for</span> item <span class="kw">in</span> nestedlist]</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>    gene_tss_list <span class="op">=</span> <span class="bu">list</span>(<span class="bu">set</span>(gene_tss_list))</span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>  output[<span class="st">'cage_predictions'</span>] <span class="op">=</span> predictions[:, <span class="dv">5110</span>] <span class="co"># a numpy array</span></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>  output[<span class="st">'gene_TSS'</span>] <span class="op">=</span> gene_tss_list <span class="co"># a list</span></span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span>(output) <span class="co"># a dictionary</span></span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> quantify_prediction_per_TSS(low_range, TSS, cage_predictions):</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true" tabindex="-1"></a>  <span class="co">'''</span></span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true" tabindex="-1"></a><span class="co">  Parameters:</span></span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true" tabindex="-1"></a><span class="co">          low_range (int): The lower interval</span></span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true" tabindex="-1"></a><span class="co">          TSS (list of integers): A list of TSS for a gene</span></span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true" tabindex="-1"></a><span class="co">          cage_predictions: A 1D numpy array or a vector of predictions from enformer corresponding to track 5110 or CAGE predictions</span></span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true" tabindex="-1"></a><span class="co">  Returns:</span></span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true" tabindex="-1"></a><span class="co">          A dictionary of gene expression predictions for each TSS for a gene</span></span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true" tabindex="-1"></a><span class="co">    '''</span></span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true" tabindex="-1"></a>  tss_predictions <span class="op">=</span> <span class="bu">dict</span>()</span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> tss <span class="kw">in</span> TSS:</span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true" tabindex="-1"></a>    bin_start <span class="op">=</span> low_range <span class="op">+</span> ((<span class="dv">768</span> <span class="op">+</span> <span class="dv">320</span>) <span class="op">*</span> <span class="dv">128</span>)</span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true" tabindex="-1"></a>    count <span class="op">=</span> <span class="op">-</span><span class="dv">1</span></span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true" tabindex="-1"></a>    <span class="cf">while</span> bin_start <span class="op">&lt;</span> tss:</span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true" tabindex="-1"></a>      bin_start <span class="op">=</span> bin_start <span class="op">+</span> <span class="dv">128</span></span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true" tabindex="-1"></a>      count <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> count <span class="op">&gt;=</span> <span class="bu">len</span>(cage_predictions)<span class="op">-</span><span class="dv">1</span>:</span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true" tabindex="-1"></a>      <span class="cf">continue</span></span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true" tabindex="-1"></a>    cage_preds <span class="op">=</span> cage_predictions[count <span class="op">-</span> <span class="dv">1</span>] <span class="op">+</span> cage_predictions[count] <span class="op">+</span> cage_predictions[count <span class="op">+</span> <span class="dv">1</span>]</span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true" tabindex="-1"></a>    tss_predictions[tss] <span class="op">=</span> cage_preds</span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span>(tss_predictions)</span>
<span id="cb7-51"><a href="#cb7-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-52"><a href="#cb7-52" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> collect_intervals(chromosomes <span class="op">=</span> [<span class="st">"22"</span>], gene_list<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb7-53"><a href="#cb7-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-54"><a href="#cb7-54" aria-hidden="true" tabindex="-1"></a>  <span class="co">'''</span></span>
<span id="cb7-55"><a href="#cb7-55" aria-hidden="true" tabindex="-1"></a><span class="co">    Parameters :</span></span>
<span id="cb7-56"><a href="#cb7-56" aria-hidden="true" tabindex="-1"></a><span class="co">      chromosomes : a list of chromosome numbers; each element should be a string format</span></span>
<span id="cb7-57"><a href="#cb7-57" aria-hidden="true" tabindex="-1"></a><span class="co">      gene_list : a list of genes; the genes should be located on those chromosomes</span></span>
<span id="cb7-58"><a href="#cb7-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-59"><a href="#cb7-59" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns :</span></span>
<span id="cb7-60"><a href="#cb7-60" aria-hidden="true" tabindex="-1"></a><span class="co">      A dictionary of genes (from gene_list) and their intervals within their respective chromosomes</span></span>
<span id="cb7-61"><a href="#cb7-61" aria-hidden="true" tabindex="-1"></a><span class="co">  '''</span></span>
<span id="cb7-62"><a href="#cb7-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-63"><a href="#cb7-63" aria-hidden="true" tabindex="-1"></a>  gene_intervals <span class="op">=</span> {} <span class="co"># Collect intervals for our genes of interest</span></span>
<span id="cb7-64"><a href="#cb7-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-65"><a href="#cb7-65" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> chrom <span class="kw">in</span> chromosomes:</span>
<span id="cb7-66"><a href="#cb7-66" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(<span class="st">"/grand/TFXcan/imlab/users/lvairus/hackenf/data/gene_chroms/gene_"</span><span class="op">+</span> chrom <span class="op">+</span> <span class="st">".txt"</span>, <span class="st">"r"</span>) <span class="im">as</span> chrom_genes:</span>
<span id="cb7-67"><a href="#cb7-67" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> line <span class="kw">in</span> chrom_genes:</span>
<span id="cb7-68"><a href="#cb7-68" aria-hidden="true" tabindex="-1"></a>        split_line <span class="op">=</span> line.strip().split(<span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb7-69"><a href="#cb7-69" aria-hidden="true" tabindex="-1"></a>        gene_intervals[split_line[<span class="dv">2</span>]] <span class="op">=</span> [</span>
<span id="cb7-70"><a href="#cb7-70" aria-hidden="true" tabindex="-1"></a>                                          split_line[<span class="dv">0</span>],</span>
<span id="cb7-71"><a href="#cb7-71" aria-hidden="true" tabindex="-1"></a>                                          <span class="bu">int</span>(split_line[<span class="dv">3</span>]),</span>
<span id="cb7-72"><a href="#cb7-72" aria-hidden="true" tabindex="-1"></a>                                          <span class="bu">int</span>(split_line[<span class="dv">4</span>])</span>
<span id="cb7-73"><a href="#cb7-73" aria-hidden="true" tabindex="-1"></a>                                        ]</span>
<span id="cb7-74"><a href="#cb7-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-75"><a href="#cb7-75" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> <span class="bu">isinstance</span>(gene_list, <span class="bu">list</span>): <span class="co"># if the user has supplied a list of genes they are interested in</span></span>
<span id="cb7-76"><a href="#cb7-76" aria-hidden="true" tabindex="-1"></a>    use_genes <span class="op">=</span> <span class="bu">dict</span>((k, gene_intervals[k]) <span class="cf">for</span> k <span class="kw">in</span> gene_list <span class="cf">if</span> k <span class="kw">in</span> gene_intervals)</span>
<span id="cb7-77"><a href="#cb7-77" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>(use_genes)</span>
<span id="cb7-78"><a href="#cb7-78" aria-hidden="true" tabindex="-1"></a>  <span class="cf">elif</span> <span class="bu">isinstance</span>(gene_list, <span class="bu">type</span>(<span class="va">None</span>)):</span>
<span id="cb7-79"><a href="#cb7-79" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>(gene_intervals)</span>
<span id="cb7-80"><a href="#cb7-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-81"><a href="#cb7-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-82"><a href="#cb7-82" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> run_predictions(gene_intervals, tss_dataframe, individuals_list<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb7-83"><a href="#cb7-83" aria-hidden="true" tabindex="-1"></a>  <span class="co">'''</span></span>
<span id="cb7-84"><a href="#cb7-84" aria-hidden="true" tabindex="-1"></a><span class="co">  Parameters :</span></span>
<span id="cb7-85"><a href="#cb7-85" aria-hidden="true" tabindex="-1"></a><span class="co">    gene_intervals : the results from calling `collect_intervals`</span></span>
<span id="cb7-86"><a href="#cb7-86" aria-hidden="true" tabindex="-1"></a><span class="co">    tss_dataframe : a list of the TSSs dataframes i.e. the TSS for the genes in the chromosomes</span></span>
<span id="cb7-87"><a href="#cb7-87" aria-hidden="true" tabindex="-1"></a><span class="co">    individuals_list : a list of individuals on which we want to make predictions; defaults to None</span></span>
<span id="cb7-88"><a href="#cb7-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-89"><a href="#cb7-89" aria-hidden="true" tabindex="-1"></a><span class="co">  Returns :</span></span>
<span id="cb7-90"><a href="#cb7-90" aria-hidden="true" tabindex="-1"></a><span class="co">    A list of predictions; the first element is the predictions around the TSS for each gene. The second is the prediction across CAGE tracks</span></span>
<span id="cb7-91"><a href="#cb7-91" aria-hidden="true" tabindex="-1"></a><span class="co">  '''</span></span>
<span id="cb7-92"><a href="#cb7-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-93"><a href="#cb7-93" aria-hidden="true" tabindex="-1"></a>  gene_output <span class="op">=</span> <span class="bu">dict</span>()</span>
<span id="cb7-94"><a href="#cb7-94" aria-hidden="true" tabindex="-1"></a>  gene_predictions <span class="op">=</span> <span class="bu">dict</span>()</span>
<span id="cb7-95"><a href="#cb7-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-96"><a href="#cb7-96" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> gene <span class="kw">in</span> gene_intervals.keys():</span>
<span id="cb7-97"><a href="#cb7-97" aria-hidden="true" tabindex="-1"></a>    gene_interval <span class="op">=</span> gene_intervals[gene]</span>
<span id="cb7-98"><a href="#cb7-98" aria-hidden="true" tabindex="-1"></a>    target_interval <span class="op">=</span> kipoiseq.Interval(<span class="st">"chr"</span> <span class="op">+</span> gene_interval[<span class="dv">0</span>],</span>
<span id="cb7-99"><a href="#cb7-99" aria-hidden="true" tabindex="-1"></a>                                        gene_interval[<span class="dv">1</span>],</span>
<span id="cb7-100"><a href="#cb7-100" aria-hidden="true" tabindex="-1"></a>                                        gene_interval[<span class="dv">2</span>]) <span class="co"># creates an interval to select the right sequences</span></span>
<span id="cb7-101"><a href="#cb7-101" aria-hidden="true" tabindex="-1"></a>    target_fa <span class="op">=</span> fasta_extractor.extract(target_interval.resize(SEQUENCE_LENGTH))  <span class="co"># extracts the fasta sequences, and resizes such that it is compatible with the sequence_length</span></span>
<span id="cb7-102"><a href="#cb7-102" aria-hidden="true" tabindex="-1"></a>    window_coords <span class="op">=</span> target_interval.resize(SEQUENCE_LENGTH) <span class="co"># we also need information about the start and end locations after resizing</span></span>
<span id="cb7-103"><a href="#cb7-103" aria-hidden="true" tabindex="-1"></a>    <span class="cf">try</span>:</span>
<span id="cb7-104"><a href="#cb7-104" aria-hidden="true" tabindex="-1"></a>      cur_gene_vars <span class="op">=</span> pd.read_csv(<span class="st">"/grand/TFXcan/imlab/users/lvairus/hackenf/data/individual_beds/chr"</span> <span class="op">+</span> gene_interval[<span class="dv">0</span>] <span class="op">+</span> <span class="st">"/chr"</span> <span class="op">+</span> gene_interval[<span class="dv">0</span>] <span class="op">+</span> <span class="st">"_"</span><span class="op">+</span> gene <span class="op">+</span> <span class="st">".bed"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, header<span class="op">=</span><span class="dv">0</span>) <span class="co"># read in the appropriate bed file for the gene</span></span>
<span id="cb7-105"><a href="#cb7-105" aria-hidden="true" tabindex="-1"></a>    <span class="cf">except</span>:</span>
<span id="cb7-106"><a href="#cb7-106" aria-hidden="true" tabindex="-1"></a>      <span class="cf">continue</span></span>
<span id="cb7-107"><a href="#cb7-107" aria-hidden="true" tabindex="-1"></a>    individual_results <span class="op">=</span> <span class="bu">dict</span>()</span>
<span id="cb7-108"><a href="#cb7-108" aria-hidden="true" tabindex="-1"></a>    individual_prediction <span class="op">=</span> <span class="bu">dict</span>()</span>
<span id="cb7-109"><a href="#cb7-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-110"><a href="#cb7-110" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">isinstance</span>(individuals_list, <span class="bu">list</span>) <span class="kw">or</span> <span class="bu">isinstance</span>(individuals_list, <span class="bu">type</span>(np.empty([<span class="dv">1</span>, <span class="dv">1</span>]))):</span>
<span id="cb7-111"><a href="#cb7-111" aria-hidden="true" tabindex="-1"></a>      use_individuals <span class="op">=</span> individuals_list</span>
<span id="cb7-112"><a href="#cb7-112" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> <span class="bu">isinstance</span>(individuals_list, <span class="bu">type</span>(<span class="va">None</span>)):</span>
<span id="cb7-113"><a href="#cb7-113" aria-hidden="true" tabindex="-1"></a>      use_individuals <span class="op">=</span> cur_gene_vars.columns[<span class="dv">4</span>:]</span>
<span id="cb7-114"><a href="#cb7-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-115"><a href="#cb7-115" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> individual <span class="kw">in</span> use_individuals:</span>
<span id="cb7-116"><a href="#cb7-116" aria-hidden="true" tabindex="-1"></a>      <span class="bu">print</span>(<span class="st">'Currently on gene </span><span class="sc">{}</span><span class="st">, and predicting on individual </span><span class="sc">{}</span><span class="st">...'</span>.<span class="bu">format</span>(gene, individual))</span>
<span id="cb7-117"><a href="#cb7-117" aria-hidden="true" tabindex="-1"></a>      <span class="co"># two haplotypes per individual</span></span>
<span id="cb7-118"><a href="#cb7-118" aria-hidden="true" tabindex="-1"></a>      haplo_1 <span class="op">=</span> <span class="bu">list</span>(target_fa[:])</span>
<span id="cb7-119"><a href="#cb7-119" aria-hidden="true" tabindex="-1"></a>      haplo_2 <span class="op">=</span> <span class="bu">list</span>(target_fa[:])</span>
<span id="cb7-120"><a href="#cb7-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-121"><a href="#cb7-121" aria-hidden="true" tabindex="-1"></a>      ref_mismatch_count <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb7-122"><a href="#cb7-122" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> i,row <span class="kw">in</span> cur_gene_vars.iterrows():</span>
<span id="cb7-123"><a href="#cb7-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-124"><a href="#cb7-124" aria-hidden="true" tabindex="-1"></a>        geno <span class="op">=</span> row[individual].split(<span class="st">"|"</span>)</span>
<span id="cb7-125"><a href="#cb7-125" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (row[<span class="st">"POS"</span>]<span class="op">-</span>window_coords.start<span class="op">-</span><span class="dv">1</span>) <span class="op">&gt;=</span> <span class="bu">len</span>(haplo_2):</span>
<span id="cb7-126"><a href="#cb7-126" aria-hidden="true" tabindex="-1"></a>          <span class="cf">continue</span></span>
<span id="cb7-127"><a href="#cb7-127" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (row[<span class="st">"POS"</span>]<span class="op">-</span>window_coords.start<span class="op">-</span><span class="dv">1</span>) <span class="op">&lt;</span> <span class="dv">0</span>:</span>
<span id="cb7-128"><a href="#cb7-128" aria-hidden="true" tabindex="-1"></a>          <span class="cf">continue</span></span>
<span id="cb7-129"><a href="#cb7-129" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> geno[<span class="dv">0</span>] <span class="op">==</span> <span class="st">"1"</span>:</span>
<span id="cb7-130"><a href="#cb7-130" aria-hidden="true" tabindex="-1"></a>          haplo_1[row[<span class="st">"POS"</span>]<span class="op">-</span>window_coords.start<span class="op">-</span><span class="dv">1</span>] <span class="op">=</span> row[<span class="st">"ALT"</span>]</span>
<span id="cb7-131"><a href="#cb7-131" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> geno[<span class="dv">1</span>] <span class="op">==</span> <span class="st">"1"</span>:</span>
<span id="cb7-132"><a href="#cb7-132" aria-hidden="true" tabindex="-1"></a>          haplo_2[row[<span class="st">"POS"</span>]<span class="op">-</span>window_coords.start<span class="op">-</span><span class="dv">1</span>] <span class="op">=</span> row[<span class="st">"ALT"</span>]</span>
<span id="cb7-133"><a href="#cb7-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-134"><a href="#cb7-134" aria-hidden="true" tabindex="-1"></a>      <span class="co"># predict on the individual's two haplotypes</span></span>
<span id="cb7-135"><a href="#cb7-135" aria-hidden="true" tabindex="-1"></a>      prediction_1 <span class="op">=</span> model.predict_on_batch(one_hot_encode(<span class="st">""</span>.join(haplo_1))[np.newaxis])[<span class="st">'human'</span>][<span class="dv">0</span>]</span>
<span id="cb7-136"><a href="#cb7-136" aria-hidden="true" tabindex="-1"></a>      prediction_2 <span class="op">=</span> model.predict_on_batch(one_hot_encode(<span class="st">""</span>.join(haplo_2))[np.newaxis])[<span class="st">'human'</span>][<span class="dv">0</span>]</span>
<span id="cb7-137"><a href="#cb7-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-138"><a href="#cb7-138" aria-hidden="true" tabindex="-1"></a>      temp_predictions <span class="op">=</span> [prediction_1[:, <span class="dv">5110</span>], prediction_2[:, <span class="dv">5110</span>]] <span class="co"># CAGE predictions we are interested in</span></span>
<span id="cb7-139"><a href="#cb7-139" aria-hidden="true" tabindex="-1"></a>      individual_prediction[individual] <span class="op">=</span> temp_predictions</span>
<span id="cb7-140"><a href="#cb7-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-141"><a href="#cb7-141" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Calculate TSS CAGE expression which correspond to column 5110 of the predictions above</span></span>
<span id="cb7-142"><a href="#cb7-142" aria-hidden="true" tabindex="-1"></a>      temp_list <span class="op">=</span> <span class="bu">list</span>()</span>
<span id="cb7-143"><a href="#cb7-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-144"><a href="#cb7-144" aria-hidden="true" tabindex="-1"></a>      pred_prepared_1 <span class="op">=</span> prepare_for_quantify_prediction_per_TSS(predictions<span class="op">=</span>prediction_1, gene<span class="op">=</span>gene, tss_df<span class="op">=</span>tss_dataframe)</span>
<span id="cb7-145"><a href="#cb7-145" aria-hidden="true" tabindex="-1"></a>      tss_predictions_1 <span class="op">=</span> quantify_prediction_per_TSS(low_range <span class="op">=</span> window_coords.start, TSS<span class="op">=</span>pred_prepared_1[<span class="st">'gene_TSS'</span>], cage_predictions<span class="op">=</span>pred_prepared_1[<span class="st">'cage_predictions'</span>])</span>
<span id="cb7-146"><a href="#cb7-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-147"><a href="#cb7-147" aria-hidden="true" tabindex="-1"></a>      pred_prepared_2 <span class="op">=</span> prepare_for_quantify_prediction_per_TSS(predictions<span class="op">=</span>prediction_2, gene<span class="op">=</span>gene, tss_df<span class="op">=</span>tss_dataframe)</span>
<span id="cb7-148"><a href="#cb7-148" aria-hidden="true" tabindex="-1"></a>      tss_predictions_2 <span class="op">=</span> quantify_prediction_per_TSS(low_range <span class="op">=</span> window_coords.start, TSS<span class="op">=</span>pred_prepared_2[<span class="st">'gene_TSS'</span>], cage_predictions<span class="op">=</span>pred_prepared_2[<span class="st">'cage_predictions'</span>])</span>
<span id="cb7-149"><a href="#cb7-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-150"><a href="#cb7-150" aria-hidden="true" tabindex="-1"></a>      temp_list.append(tss_predictions_1)</span>
<span id="cb7-151"><a href="#cb7-151" aria-hidden="true" tabindex="-1"></a>      temp_list.append(tss_predictions_2) <span class="co"># results here are a dictionary for each TSS for each haplotype</span></span>
<span id="cb7-152"><a href="#cb7-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-153"><a href="#cb7-153" aria-hidden="true" tabindex="-1"></a>      individual_results[individual] <span class="op">=</span> temp_list <span class="co"># save for the individual</span></span>
<span id="cb7-154"><a href="#cb7-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-155"><a href="#cb7-155" aria-hidden="true" tabindex="-1"></a>    gene_output[gene] <span class="op">=</span> individual_results</span>
<span id="cb7-156"><a href="#cb7-156" aria-hidden="true" tabindex="-1"></a>    gene_predictions[gene] <span class="op">=</span> individual_prediction</span>
<span id="cb7-157"><a href="#cb7-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-158"><a href="#cb7-158" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span>([gene_output, gene_predictions])</span>
<span id="cb7-159"><a href="#cb7-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-160"><a href="#cb7-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-161"><a href="#cb7-161" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> collect_target_intervals(gene_intervals):</span>
<span id="cb7-162"><a href="#cb7-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-163"><a href="#cb7-163" aria-hidden="true" tabindex="-1"></a>  <span class="co">'''</span></span>
<span id="cb7-164"><a href="#cb7-164" aria-hidden="true" tabindex="-1"></a><span class="co">  Returns a dictionary of Interval objects (from kipoiseq) for each gene corresponding to the locations of the gene</span></span>
<span id="cb7-165"><a href="#cb7-165" aria-hidden="true" tabindex="-1"></a><span class="co">  '''</span></span>
<span id="cb7-166"><a href="#cb7-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-167"><a href="#cb7-167" aria-hidden="true" tabindex="-1"></a>  target_intervals_dict <span class="op">=</span> <span class="bu">dict</span>()</span>
<span id="cb7-168"><a href="#cb7-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-169"><a href="#cb7-169" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> gene <span class="kw">in</span> gene_intervals.keys():</span>
<span id="cb7-170"><a href="#cb7-170" aria-hidden="true" tabindex="-1"></a>    gene_interval <span class="op">=</span> gene_intervals[gene]</span>
<span id="cb7-171"><a href="#cb7-171" aria-hidden="true" tabindex="-1"></a>    target_interval <span class="op">=</span> kipoiseq.Interval(<span class="st">"chr"</span> <span class="op">+</span> gene_interval[<span class="dv">0</span>],</span>
<span id="cb7-172"><a href="#cb7-172" aria-hidden="true" tabindex="-1"></a>                                        gene_interval[<span class="dv">1</span>],</span>
<span id="cb7-173"><a href="#cb7-173" aria-hidden="true" tabindex="-1"></a>                                        gene_interval[<span class="dv">2</span>])</span>
<span id="cb7-174"><a href="#cb7-174" aria-hidden="true" tabindex="-1"></a>    target_intervals_dict[gene] <span class="op">=</span> target_interval</span>
<span id="cb7-175"><a href="#cb7-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-176"><a href="#cb7-176" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span>(target_intervals_dict)</span>
<span id="cb7-177"><a href="#cb7-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-178"><a href="#cb7-178" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> prepare_for_plot_tracks(gene, individual, all_predictions, chromosome<span class="op">=</span>[<span class="st">'22'</span>]):</span>
<span id="cb7-179"><a href="#cb7-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-180"><a href="#cb7-180" aria-hidden="true" tabindex="-1"></a>  <span class="co">'''</span></span>
<span id="cb7-181"><a href="#cb7-181" aria-hidden="true" tabindex="-1"></a><span class="co">  This returns a dictionary of gene tracks and gene intervals, prepared for the function plot_tracks.</span></span>
<span id="cb7-182"><a href="#cb7-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-183"><a href="#cb7-183" aria-hidden="true" tabindex="-1"></a><span class="co">  Parameters:</span></span>
<span id="cb7-184"><a href="#cb7-184" aria-hidden="true" tabindex="-1"></a><span class="co">    - gene</span></span>
<span id="cb7-185"><a href="#cb7-185" aria-hidden="true" tabindex="-1"></a><span class="co">    - individual</span></span>
<span id="cb7-186"><a href="#cb7-186" aria-hidden="true" tabindex="-1"></a><span class="co">    - all_predictions</span></span>
<span id="cb7-187"><a href="#cb7-187" aria-hidden="true" tabindex="-1"></a><span class="co">  '''</span></span>
<span id="cb7-188"><a href="#cb7-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-189"><a href="#cb7-189" aria-hidden="true" tabindex="-1"></a>  haplo_predictions <span class="op">=</span> all_predictions[gene][individual]</span>
<span id="cb7-190"><a href="#cb7-190" aria-hidden="true" tabindex="-1"></a>  gene_tracks <span class="op">=</span> {gene <span class="op">+</span> <span class="st">' | '</span> <span class="op">+</span> individual <span class="op">+</span> <span class="st">' | haplotype 1'</span>: np.log10(<span class="dv">1</span> <span class="op">+</span> haplo_predictions[<span class="dv">0</span>]),</span>
<span id="cb7-191"><a href="#cb7-191" aria-hidden="true" tabindex="-1"></a>                gene <span class="op">+</span> <span class="st">' | '</span> <span class="op">+</span> individual <span class="op">+</span> <span class="st">' | haplotype 2'</span>: np.log10(<span class="dv">1</span> <span class="op">+</span> haplo_predictions[<span class="dv">1</span>])}</span>
<span id="cb7-192"><a href="#cb7-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-193"><a href="#cb7-193" aria-hidden="true" tabindex="-1"></a>  gene_intervals <span class="op">=</span> collect_intervals(chromosomes<span class="op">=</span>chromosome, gene_list<span class="op">=</span>[gene])</span>
<span id="cb7-194"><a href="#cb7-194" aria-hidden="true" tabindex="-1"></a>  gene_intervals <span class="op">=</span> collect_target_intervals(gene_intervals)</span>
<span id="cb7-195"><a href="#cb7-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-196"><a href="#cb7-196" aria-hidden="true" tabindex="-1"></a>  output <span class="op">=</span> <span class="bu">dict</span>()</span>
<span id="cb7-197"><a href="#cb7-197" aria-hidden="true" tabindex="-1"></a>  output[<span class="st">'gene_tracks'</span>] <span class="op">=</span> gene_tracks</span>
<span id="cb7-198"><a href="#cb7-198" aria-hidden="true" tabindex="-1"></a>  output[<span class="st">'gene_intervals'</span>] <span class="op">=</span> gene_intervals[gene]</span>
<span id="cb7-199"><a href="#cb7-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-200"><a href="#cb7-200" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span>(output)</span>
<span id="cb7-201"><a href="#cb7-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-202"><a href="#cb7-202" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> check_individuals(path_to_bed_file, list_of_individuals):</span>
<span id="cb7-203"><a href="#cb7-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-204"><a href="#cb7-204" aria-hidden="true" tabindex="-1"></a>  <span class="co">'''</span></span>
<span id="cb7-205"><a href="#cb7-205" aria-hidden="true" tabindex="-1"></a><span class="co">  Checks if an individual is missing in bed variation files.</span></span>
<span id="cb7-206"><a href="#cb7-206" aria-hidden="true" tabindex="-1"></a><span class="co">  These individuals should be removed prior to training</span></span>
<span id="cb7-207"><a href="#cb7-207" aria-hidden="true" tabindex="-1"></a><span class="co">  '''</span></span>
<span id="cb7-208"><a href="#cb7-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-209"><a href="#cb7-209" aria-hidden="true" tabindex="-1"></a>  myfile <span class="op">=</span> <span class="bu">open</span>(path_to_bed_file, <span class="st">'r'</span>)</span>
<span id="cb7-210"><a href="#cb7-210" aria-hidden="true" tabindex="-1"></a>  myline <span class="op">=</span> myfile.readline()</span>
<span id="cb7-211"><a href="#cb7-211" aria-hidden="true" tabindex="-1"></a>  bed_names <span class="op">=</span> myline.split(<span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)[<span class="dv">4</span>:]</span>
<span id="cb7-212"><a href="#cb7-212" aria-hidden="true" tabindex="-1"></a>  myfile.close()</span>
<span id="cb7-213"><a href="#cb7-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-214"><a href="#cb7-214" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> <span class="bu">set</span>(list_of_individuals).issubset(<span class="bu">set</span>(bed_names)) <span class="op">==</span> <span class="va">False</span>:</span>
<span id="cb7-215"><a href="#cb7-215" aria-hidden="true" tabindex="-1"></a>    missing <span class="op">=</span> <span class="bu">list</span>(<span class="bu">set</span>(list_of_individuals).difference(bed_names))</span>
<span id="cb7-216"><a href="#cb7-216" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">'This (or these) individual(s) is/are not present: </span><span class="sc">{}</span><span class="st">'</span>.<span class="bu">format</span>(missing))</span>
<span id="cb7-217"><a href="#cb7-217" aria-hidden="true" tabindex="-1"></a>  <span class="cf">else</span>:</span>
<span id="cb7-218"><a href="#cb7-218" aria-hidden="true" tabindex="-1"></a>    missing <span class="op">=</span> []</span>
<span id="cb7-219"><a href="#cb7-219" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">'All individuals are present in the bed file.'</span>)</span>
<span id="cb7-220"><a href="#cb7-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-221"><a href="#cb7-221" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span>(missing)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> geno_to_seq(gene, individual):</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>      <span class="co"># two haplotypes per individual</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  haplo_1 <span class="op">=</span> <span class="bu">list</span>(target_fa[:])</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  haplo_2 <span class="op">=</span> <span class="bu">list</span>(target_fa[:])</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  ref_mismatch_count <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> i,row <span class="kw">in</span> cur_gene_vars.iterrows():</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    geno <span class="op">=</span> row[individual].split(<span class="st">"|"</span>)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (row[<span class="st">"POS"</span>]<span class="op">-</span>window_coords.start<span class="op">-</span><span class="dv">1</span>) <span class="op">&gt;=</span> <span class="bu">len</span>(haplo_2):</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>      <span class="cf">continue</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (row[<span class="st">"POS"</span>]<span class="op">-</span>window_coords.start<span class="op">-</span><span class="dv">1</span>) <span class="op">&lt;</span> <span class="dv">0</span>:</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>      <span class="cf">continue</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> geno[<span class="dv">0</span>] <span class="op">==</span> <span class="st">"1"</span>:</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>      haplo_1[row[<span class="st">"POS"</span>]<span class="op">-</span>window_coords.start<span class="op">-</span><span class="dv">1</span>] <span class="op">=</span> row[<span class="st">"ALT"</span>]</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> geno[<span class="dv">1</span>] <span class="op">==</span> <span class="st">"1"</span>:</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>      haplo_2[row[<span class="st">"POS"</span>]<span class="op">-</span>window_coords.start<span class="op">-</span><span class="dv">1</span>] <span class="op">=</span> row[<span class="st">"ALT"</span>]</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> haplo_1, haplo_2</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>      <span class="co"># predict on the individual's two haplotypes</span></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>    </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> run_predictions2(gene, chrom, indiv):</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">'''</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Parameters :</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co">       gene: gene to run (string)</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co">       chrom: chrom gene is on (string)</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co">       indiv: individual to run (string)</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co">       chrNtss: variable of tss for chr</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns :</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co">        A list of predictions; the first element is the predictions around the TSS for each gene. The second is the prediction across CAGE tracks</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co">    '''</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>    gene_intervals <span class="op">=</span> collect_intervals(chromosomes<span class="op">=</span>[chrom], gene_list<span class="op">=</span>[gene])</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>    individuals_list <span class="op">=</span> [indiv]</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> gene <span class="kw">in</span> gene_intervals.keys():</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>        <span class="kw">global</span> fasta_extractor</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>        gene_interval <span class="op">=</span> gene_intervals[gene]</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>        target_interval <span class="op">=</span> kipoiseq.Interval(<span class="st">"chr"</span> <span class="op">+</span> gene_interval[<span class="dv">0</span>],</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>                                            gene_interval[<span class="dv">1</span>],</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>                                            gene_interval[<span class="dv">2</span>]) <span class="co"># creates an interval to select the right sequences</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>        target_fa <span class="op">=</span> fasta_extractor.extract(target_interval.resize(SEQUENCE_LENGTH))  <span class="co"># extracts the fasta sequences, and resizes such that it is compatible with the sequence_length</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>        window_coords <span class="op">=</span> target_interval.resize(SEQUENCE_LENGTH) <span class="co"># we also need information about the start and end locations after resizing</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>        <span class="cf">try</span>:</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>            cur_gene_vars <span class="op">=</span> pd.read_csv(<span class="st">"/grand/TFXcan/imlab/users/lvairus/hackenf/data/individual_beds/chr"</span> <span class="op">+</span> gene_interval[<span class="dv">0</span>] <span class="op">+</span> <span class="st">"/chr"</span> <span class="op">+</span> gene_interval[<span class="dv">0</span>] <span class="op">+</span> <span class="st">"_"</span><span class="op">+</span> gene <span class="op">+</span> <span class="st">".bed"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, header<span class="op">=</span><span class="dv">0</span>) <span class="co"># read in the appropriate bed file for the gene</span></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>        <span class="cf">except</span>:</span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>            <span class="cf">continue</span></span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>        individual_prediction <span class="op">=</span> <span class="bu">dict</span>()</span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">isinstance</span>(individuals_list, <span class="bu">list</span>) <span class="kw">or</span> <span class="bu">isinstance</span>(individuals_list, <span class="bu">type</span>(np.empty([<span class="dv">1</span>, <span class="dv">1</span>]))):</span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>            use_individuals <span class="op">=</span> individuals_list</span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elif</span> <span class="bu">isinstance</span>(individuals_list, <span class="bu">type</span>(<span class="va">None</span>)):</span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a>            use_individuals <span class="op">=</span> cur_gene_vars.columns[<span class="dv">4</span>:]</span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-34"><a href="#cb9-34" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> individual <span class="kw">in</span> use_individuals:</span>
<span id="cb9-35"><a href="#cb9-35" aria-hidden="true" tabindex="-1"></a>            <span class="co"># two haplotypes per individual</span></span>
<span id="cb9-36"><a href="#cb9-36" aria-hidden="true" tabindex="-1"></a>            haplo_1 <span class="op">=</span> <span class="bu">list</span>(target_fa[:])</span>
<span id="cb9-37"><a href="#cb9-37" aria-hidden="true" tabindex="-1"></a>            haplo_2 <span class="op">=</span> <span class="bu">list</span>(target_fa[:])</span>
<span id="cb9-38"><a href="#cb9-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-39"><a href="#cb9-39" aria-hidden="true" tabindex="-1"></a>            ref_mismatch_count <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb9-40"><a href="#cb9-40" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> i,row <span class="kw">in</span> cur_gene_vars.iterrows():</span>
<span id="cb9-41"><a href="#cb9-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-42"><a href="#cb9-42" aria-hidden="true" tabindex="-1"></a>                geno <span class="op">=</span> row[individual].split(<span class="st">"|"</span>)</span>
<span id="cb9-43"><a href="#cb9-43" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> (row[<span class="st">"POS"</span>]<span class="op">-</span>window_coords.start<span class="op">-</span><span class="dv">1</span>) <span class="op">&gt;=</span> <span class="bu">len</span>(haplo_2):</span>
<span id="cb9-44"><a href="#cb9-44" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">continue</span></span>
<span id="cb9-45"><a href="#cb9-45" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> (row[<span class="st">"POS"</span>]<span class="op">-</span>window_coords.start<span class="op">-</span><span class="dv">1</span>) <span class="op">&lt;</span> <span class="dv">0</span>:</span>
<span id="cb9-46"><a href="#cb9-46" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">continue</span></span>
<span id="cb9-47"><a href="#cb9-47" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> geno[<span class="dv">0</span>] <span class="op">==</span> <span class="st">"1"</span>:</span>
<span id="cb9-48"><a href="#cb9-48" aria-hidden="true" tabindex="-1"></a>                    haplo_1[row[<span class="st">"POS"</span>]<span class="op">-</span>window_coords.start<span class="op">-</span><span class="dv">1</span>] <span class="op">=</span> row[<span class="st">"ALT"</span>]</span>
<span id="cb9-49"><a href="#cb9-49" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> geno[<span class="dv">1</span>] <span class="op">==</span> <span class="st">"1"</span>:</span>
<span id="cb9-50"><a href="#cb9-50" aria-hidden="true" tabindex="-1"></a>                    haplo_2[row[<span class="st">"POS"</span>]<span class="op">-</span>window_coords.start<span class="op">-</span><span class="dv">1</span>] <span class="op">=</span> row[<span class="st">"ALT"</span>]</span>
<span id="cb9-51"><a href="#cb9-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-52"><a href="#cb9-52" aria-hidden="true" tabindex="-1"></a>            <span class="co"># predict on the individual's two haplotypes</span></span>
<span id="cb9-53"><a href="#cb9-53" aria-hidden="true" tabindex="-1"></a>            <span class="kw">global</span> model</span>
<span id="cb9-54"><a href="#cb9-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-55"><a href="#cb9-55" aria-hidden="true" tabindex="-1"></a>            ohe_haplo_1 <span class="op">=</span> one_hot_encode(<span class="st">""</span>.join(haplo_1))[np.newaxis]</span>
<span id="cb9-56"><a href="#cb9-56" aria-hidden="true" tabindex="-1"></a>            ohe_haplo_2 <span class="op">=</span> one_hot_encode(<span class="st">""</span>.join(haplo_2))[np.newaxis]</span>
<span id="cb9-57"><a href="#cb9-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-58"><a href="#cb9-58" aria-hidden="true" tabindex="-1"></a>            ohe_haplo_avg <span class="op">=</span> np.add(ohe_haplo_1, ohe_haplo_2) <span class="op">/</span> <span class="dv">2</span></span>
<span id="cb9-59"><a href="#cb9-59" aria-hidden="true" tabindex="-1"></a>                                         </span>
<span id="cb9-60"><a href="#cb9-60" aria-hidden="true" tabindex="-1"></a>            prediction_1 <span class="op">=</span> model.predict_on_batch(ohe_haplo_1)[<span class="st">'human'</span>][<span class="dv">0</span>]</span>
<span id="cb9-61"><a href="#cb9-61" aria-hidden="true" tabindex="-1"></a>            prediction_2 <span class="op">=</span> model.predict_on_batch(ohe_haplo_2)[<span class="st">'human'</span>][<span class="dv">0</span>]</span>
<span id="cb9-62"><a href="#cb9-62" aria-hidden="true" tabindex="-1"></a>            prediction_avg <span class="op">=</span> model.predict_on_batch(ohe_haplo_avg)[<span class="st">'human'</span>][<span class="dv">0</span>]</span>
<span id="cb9-63"><a href="#cb9-63" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb9-64"><a href="#cb9-64" aria-hidden="true" tabindex="-1"></a>            post_avg <span class="op">=</span> (prediction_1 <span class="op">+</span> prediction_2) <span class="op">/</span> <span class="dv">2</span></span>
<span id="cb9-65"><a href="#cb9-65" aria-hidden="true" tabindex="-1"></a>            pre_avg <span class="op">=</span> prediction_avg</span>
<span id="cb9-66"><a href="#cb9-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-67"><a href="#cb9-67" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>([pre_avg, post_avg])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_summary(arr):</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>    stats <span class="op">=</span> {</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>        <span class="st">"mean"</span>: np.mean(arr),</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">"median"</span>: np.median(arr),</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">"std_dev"</span>: np.std(arr),</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"minimum"</span>: np.<span class="bu">min</span>(arr),</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">"maximum"</span>: np.<span class="bu">max</span>(arr),</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>        <span class="st">"total_sum"</span>: np.<span class="bu">sum</span>(arr),</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>        <span class="st">"q1"</span>: np.percentile(arr, <span class="dv">25</span>),</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>        <span class="st">"q2"</span>: np.percentile(arr, <span class="dv">50</span>),</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>        <span class="st">"q3"</span>: np.percentile(arr, <span class="dv">75</span>),</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>        <span class="st">"iqr"</span>: np.percentile(arr, <span class="dv">75</span>) <span class="op">-</span> np.percentile(arr, <span class="dv">25</span>)</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> stats</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> make_hist(arr, bin_num):</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>    plt.hist(arr, bins<span class="op">=</span>bin_num)</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Add labels and title</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>    plt.xlabel(<span class="st">'Value'</span>)</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>    plt.ylabel(<span class="st">'Frequency'</span>)</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Display the plot</span></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_relmax(arr):</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    col_max <span class="op">=</span> np.<span class="bu">max</span>(arr, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    relmax <span class="op">=</span> arr <span class="op">/</span> col_max</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> relmax</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_relmed(arr):</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>    col_med <span class="op">=</span> np.median(arr, axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    relmed <span class="op">=</span> arr <span class="op">/</span> col_med</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> relmed</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> My_class:</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, gene, chrom, indiv):</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>        <span class="co"># save starting variables</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.gene <span class="op">=</span> gene</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.chrom <span class="op">=</span> chrom</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.indiv <span class="op">=</span> indiv</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>        <span class="co"># run predictions</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.pre_avg, <span class="va">self</span>.post_avg <span class="op">=</span> run_predictions2(gene, chrom, indiv)</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>        <span class="co"># get difference matrices</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.diffmat <span class="op">=</span> <span class="va">self</span>.pre_avg <span class="op">-</span> <span class="va">self</span>.post_avg</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.abs_diffmat <span class="op">=</span> <span class="bu">abs</span>(<span class="va">self</span>.diffmat) <span class="co"># check this</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.rel_diffmat <span class="op">=</span> <span class="va">self</span>.abs_diffmat <span class="op">/</span> (<span class="bu">abs</span>(<span class="va">self</span>.pre_avg) <span class="op">+</span> <span class="bu">abs</span>(<span class="va">self</span>.post_avg) <span class="op">+</span> <span class="dv">1</span><span class="op">**-</span><span class="dv">16</span>) <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.relmax_diffmat <span class="op">=</span> get_relmax(<span class="va">self</span>.abs_diffmat)</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.relmed_diffmat <span class="op">=</span> get_relmed(<span class="va">self</span>.abs_diffmat)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> get_diffmats(<span class="va">self</span>):</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> [<span class="va">self</span>.abs_diffmat, <span class="va">self</span>.rel_diffmat, <span class="va">self</span>.relmax_diffmat, <span class="va">self</span>.relmed_diffmat]</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> get_summary(<span class="va">self</span>, arr):</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>        stats <span class="op">=</span> {</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>        <span class="st">"mean"</span>: np.mean(arr),</span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>        <span class="st">"median"</span>: np.median(arr),</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>        <span class="st">"std_dev"</span>: np.std(arr),</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a>        <span class="st">"minimum"</span>: np.<span class="bu">min</span>(arr),</span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>        <span class="st">"maximum"</span>: np.<span class="bu">max</span>(arr),</span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>        <span class="st">"total_sum"</span>: np.<span class="bu">sum</span>(arr),</span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>        <span class="st">"q1"</span>: np.percentile(arr, <span class="dv">25</span>),</span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>        <span class="st">"q2"</span>: np.percentile(arr, <span class="dv">50</span>),</span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a>        <span class="st">"q3"</span>: np.percentile(arr, <span class="dv">75</span>),</span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>        <span class="st">"iqr"</span>: np.percentile(arr, <span class="dv">75</span>) <span class="op">-</span> np.percentile(arr, <span class="dv">25</span>)</span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> stats</span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> get_outliers(<span class="va">self</span>, arr, tol):</span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true" tabindex="-1"></a>        outliers <span class="op">=</span> arr[arr<span class="op">&gt;</span>tol]</span>
<span id="cb13-36"><a href="#cb13-36" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> outliers</span>
<span id="cb13-37"><a href="#cb13-37" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-38"><a href="#cb13-38" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> make_hist(<span class="va">self</span>, arr, bin_num):</span>
<span id="cb13-39"><a href="#cb13-39" aria-hidden="true" tabindex="-1"></a>        plt.hist(arr, bins<span class="op">=</span>bin_num)</span>
<span id="cb13-40"><a href="#cb13-40" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Add labels and title</span></span>
<span id="cb13-41"><a href="#cb13-41" aria-hidden="true" tabindex="-1"></a>        plt.xlabel(<span class="st">'Value'</span>)</span>
<span id="cb13-42"><a href="#cb13-42" aria-hidden="true" tabindex="-1"></a>        plt.ylabel(<span class="st">'Frequency'</span>)</span>
<span id="cb13-43"><a href="#cb13-43" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Display the plot</span></span>
<span id="cb13-44"><a href="#cb13-44" aria-hidden="true" tabindex="-1"></a>        plt.show()</span>
<span id="cb13-45"><a href="#cb13-45" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="va">None</span></span>
<span id="cb13-46"><a href="#cb13-46" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-47"><a href="#cb13-47" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> get_outlier_inds(<span class="va">self</span>, arr, tol):</span>
<span id="cb13-48"><a href="#cb13-48" aria-hidden="true" tabindex="-1"></a>        <span class="co"># get inds of outliers list of [row,col] lists</span></span>
<span id="cb13-49"><a href="#cb13-49" aria-hidden="true" tabindex="-1"></a>        outlier_inds <span class="op">=</span> np.where(arr <span class="op">&gt;</span> tol)</span>
<span id="cb13-50"><a href="#cb13-50" aria-hidden="true" tabindex="-1"></a>        outlier_inds_arr <span class="op">=</span> np.array(<span class="bu">list</span>(<span class="bu">zip</span>(outlier_inds[<span class="dv">0</span>], outlier_inds[<span class="dv">1</span>])))</span>
<span id="cb13-51"><a href="#cb13-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-52"><a href="#cb13-52" aria-hidden="true" tabindex="-1"></a>        <span class="co"># define outlier row inds</span></span>
<span id="cb13-53"><a href="#cb13-53" aria-hidden="true" tabindex="-1"></a>        row_inds <span class="op">=</span> outlier_inds[<span class="dv">0</span>]      <span class="co"># all outlier row inds</span></span>
<span id="cb13-54"><a href="#cb13-54" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb13-55"><a href="#cb13-55" aria-hidden="true" tabindex="-1"></a>        <span class="co">#define outlier col inds</span></span>
<span id="cb13-56"><a href="#cb13-56" aria-hidden="true" tabindex="-1"></a>        col_inds <span class="op">=</span> outlier_inds[<span class="dv">1</span>]      <span class="co"># all outlier col inds</span></span>
<span id="cb13-57"><a href="#cb13-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-58"><a href="#cb13-58" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"to get list of unique inds: set(inds_arr) "</span>)</span>
<span id="cb13-59"><a href="#cb13-59" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"get count of inds: Counter(inds_arr)"</span>)</span>
<span id="cb13-60"><a href="#cb13-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-61"><a href="#cb13-61" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> (outlier_inds_arr, row_inds, col_inds)</span>
<span id="cb13-62"><a href="#cb13-62" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-63"><a href="#cb13-63" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> make_table(<span class="va">self</span>, arr, tol):</span>
<span id="cb13-64"><a href="#cb13-64" aria-hidden="true" tabindex="-1"></a>        <span class="co"># input difference arr and tolerance</span></span>
<span id="cb13-65"><a href="#cb13-65" aria-hidden="true" tabindex="-1"></a>        <span class="co"># get outliers with tolerance</span></span>
<span id="cb13-66"><a href="#cb13-66" aria-hidden="true" tabindex="-1"></a>        outs <span class="op">=</span> arr[arr<span class="op">&gt;</span>tol]</span>
<span id="cb13-67"><a href="#cb13-67" aria-hidden="true" tabindex="-1"></a>        out_inds <span class="op">=</span> np.where(arr <span class="op">&gt;</span> tol)</span>
<span id="cb13-68"><a href="#cb13-68" aria-hidden="true" tabindex="-1"></a>        col_inds <span class="op">=</span> out_inds[<span class="dv">1</span>]   </span>
<span id="cb13-69"><a href="#cb13-69" aria-hidden="true" tabindex="-1"></a>        row_inds <span class="op">=</span> out_inds[<span class="dv">0</span>]</span>
<span id="cb13-70"><a href="#cb13-70" aria-hidden="true" tabindex="-1"></a>        <span class="co"># labels for track names from col names</span></span>
<span id="cb13-71"><a href="#cb13-71" aria-hidden="true" tabindex="-1"></a>        <span class="kw">global</span> track_names</span>
<span id="cb13-72"><a href="#cb13-72" aria-hidden="true" tabindex="-1"></a>        <span class="co"># row_inds </span></span>
<span id="cb13-73"><a href="#cb13-73" aria-hidden="true" tabindex="-1"></a>        <span class="co"># outlier values </span></span>
<span id="cb13-74"><a href="#cb13-74" aria-hidden="true" tabindex="-1"></a>        data <span class="op">=</span> {</span>
<span id="cb13-75"><a href="#cb13-75" aria-hidden="true" tabindex="-1"></a>            <span class="st">'track'</span>: [track_names[i] <span class="cf">for</span> i <span class="kw">in</span> col_inds],</span>
<span id="cb13-76"><a href="#cb13-76" aria-hidden="true" tabindex="-1"></a>            <span class="st">'bin'</span>: row_inds,</span>
<span id="cb13-77"><a href="#cb13-77" aria-hidden="true" tabindex="-1"></a>            <span class="st">'diff'</span>: outs</span>
<span id="cb13-78"><a href="#cb13-78" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb13-79"><a href="#cb13-79" aria-hidden="true" tabindex="-1"></a>        df <span class="op">=</span> pd.DataFrame(data)</span>
<span id="cb13-80"><a href="#cb13-80" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> df</span>
<span id="cb13-81"><a href="#cb13-81" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-82"><a href="#cb13-82" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> plot_tol_numouts(<span class="va">self</span>, arr, mintol, maxtol, steps):</span>
<span id="cb13-83"><a href="#cb13-83" aria-hidden="true" tabindex="-1"></a>        x <span class="op">=</span> np.linspace(maxtol, mintol, steps)</span>
<span id="cb13-84"><a href="#cb13-84" aria-hidden="true" tabindex="-1"></a>        y <span class="op">=</span> []</span>
<span id="cb13-85"><a href="#cb13-85" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> tol <span class="kw">in</span> x:</span>
<span id="cb13-86"><a href="#cb13-86" aria-hidden="true" tabindex="-1"></a>            y.append(<span class="bu">len</span>(<span class="va">self</span>.get_outliers(arr, tol)))</span>
<span id="cb13-87"><a href="#cb13-87" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Create scatter plot</span></span>
<span id="cb13-88"><a href="#cb13-88" aria-hidden="true" tabindex="-1"></a>        plt.scatter(x, y)</span>
<span id="cb13-89"><a href="#cb13-89" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Add labels and title</span></span>
<span id="cb13-90"><a href="#cb13-90" aria-hidden="true" tabindex="-1"></a>        plt.xlabel(<span class="st">'Tolerance'</span>)</span>
<span id="cb13-91"><a href="#cb13-91" aria-hidden="true" tabindex="-1"></a>        plt.ylabel(<span class="st">'Amount of Outliers'</span>)</span>
<span id="cb13-92"><a href="#cb13-92" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Display the plot</span></span>
<span id="cb13-93"><a href="#cb13-93" aria-hidden="true" tabindex="-1"></a>        plt.show()</span>
<span id="cb13-94"><a href="#cb13-94" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="va">None</span></span>
<span id="cb13-95"><a href="#cb13-95" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> print_comparisons(pre_average, post_average, tolerance):</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">dict</span> <span class="op">=</span> {}</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    diffmat <span class="op">=</span> pre_average <span class="op">-</span> post_average</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># abs diff matrix</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">abs</span> <span class="op">=</span> {}</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Absolute Difference:"</span>)</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># make abs diff mat</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>    abs_mat <span class="op">=</span> np.sqrt(np.square(diffmat))</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>    <span class="bu">abs</span>[<span class="st">'mat'</span>] <span class="op">=</span> abs_mat</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get summary of mat</span></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>    abs_mat_summary <span class="op">=</span> get_summary(abs_mat)</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>    <span class="bu">abs</span>[<span class="st">'mat_summary'</span>] <span class="op">=</span> abs_mat_summary</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Summary Statistics: </span><span class="ch">\n</span><span class="sc">{</span>abs_mat_summary<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get outlier differences</span></span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>    outlier_abs <span class="op">=</span> abs_mat[abs_mat<span class="op">&gt;</span>tolerance]</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>    <span class="bu">abs</span>[<span class="st">"outliers"</span>] <span class="op">=</span> outlier_abs</span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'num outliers: </span><span class="ch">\n</span><span class="sc">{</span><span class="bu">len</span>(outlier_abs)<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a>    <span class="co"># summary of outliers</span></span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a>    outlier_abs_summary <span class="op">=</span> get_summary(outlier_abs)</span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a>    <span class="bu">abs</span>[<span class="st">'outliers-summary'</span>] <span class="op">=</span> outlier_abs_summary</span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Summary of Outliers: </span><span class="ch">\n</span><span class="sc">{</span>outlier_abs_summary<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true" tabindex="-1"></a>    <span class="co"># histogram of outliers </span></span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Histogram of Outliers for Absolute Difference"</span>)</span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true" tabindex="-1"></a>    make_hist(outlier_abs, <span class="dv">20</span>)</span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get inds of outliers list of [row,col] lists</span></span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true" tabindex="-1"></a>    outlier_abs_inds <span class="op">=</span> np.where(abs_mat <span class="op">&gt;</span> tolerance)</span>
<span id="cb14-35"><a href="#cb14-35" aria-hidden="true" tabindex="-1"></a>    outlier_abs_inds_arr <span class="op">=</span> np.array(<span class="bu">list</span>(<span class="bu">zip</span>(outlier_abs_inds[<span class="dv">0</span>], outlier_abs_inds[<span class="dv">1</span>])))</span>
<span id="cb14-36"><a href="#cb14-36" aria-hidden="true" tabindex="-1"></a>    <span class="bu">abs</span>[<span class="st">'outlier-inds'</span>] <span class="op">=</span> outlier_abs_inds_arr</span>
<span id="cb14-37"><a href="#cb14-37" aria-hidden="true" tabindex="-1"></a>    <span class="co"># define outlier row inds</span></span>
<span id="cb14-38"><a href="#cb14-38" aria-hidden="true" tabindex="-1"></a>    row_inds <span class="op">=</span> outlier_abs_inds[<span class="dv">0</span>]      <span class="co"># all outlier row inds</span></span>
<span id="cb14-39"><a href="#cb14-39" aria-hidden="true" tabindex="-1"></a>    row_inds_uq <span class="op">=</span> <span class="bu">set</span>(row_inds)         <span class="co"># unique outlier row inds</span></span>
<span id="cb14-40"><a href="#cb14-40" aria-hidden="true" tabindex="-1"></a>    row_inds_count <span class="op">=</span> Counter(row_inds)  <span class="co"># counter of outlier row inds</span></span>
<span id="cb14-41"><a href="#cb14-41" aria-hidden="true" tabindex="-1"></a>    <span class="co">#define outlier col inds</span></span>
<span id="cb14-42"><a href="#cb14-42" aria-hidden="true" tabindex="-1"></a>    col_inds <span class="op">=</span> outlier_abs_inds[<span class="dv">1</span>]      <span class="co"># all outlier col inds</span></span>
<span id="cb14-43"><a href="#cb14-43" aria-hidden="true" tabindex="-1"></a>    col_inds_uq <span class="op">=</span> <span class="bu">set</span>(col_inds)         <span class="co"># unique outlier col inds</span></span>
<span id="cb14-44"><a href="#cb14-44" aria-hidden="true" tabindex="-1"></a>    col_inds_count <span class="op">=</span> Counter(col_inds)  <span class="co"># counter of outlier col inds</span></span>
<span id="cb14-45"><a href="#cb14-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-46"><a href="#cb14-46" aria-hidden="true" tabindex="-1"></a>    <span class="co">#hist of row outlier inds</span></span>
<span id="cb14-47"><a href="#cb14-47" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'num unique row outlier inds: </span><span class="sc">{</span><span class="bu">len</span>(row_inds_uq)<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb14-48"><a href="#cb14-48" aria-hidden="true" tabindex="-1"></a>    make_hist(row_inds)</span>
<span id="cb14-49"><a href="#cb14-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-50"><a href="#cb14-50" aria-hidden="true" tabindex="-1"></a>    <span class="co">#hist of col outlier inds</span></span>
<span id="cb14-51"><a href="#cb14-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-52"><a href="#cb14-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-53"><a href="#cb14-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-54"><a href="#cb14-54" aria-hidden="true" tabindex="-1"></a>    <span class="co"># rel diff matrix</span></span>
<span id="cb14-55"><a href="#cb14-55" aria-hidden="true" tabindex="-1"></a>    <span class="co">#rel_diff = (abs_diff) / ((pre_average + post_average) + 10**-16)</span></span>
<span id="cb14-56"><a href="#cb14-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-57"><a href="#cb14-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-58"><a href="#cb14-58" aria-hidden="true" tabindex="-1"></a>    <span class="co"># rel max diff matrix</span></span>
<span id="cb14-59"><a href="#cb14-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-60"><a href="#cb14-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-61"><a href="#cb14-61" aria-hidden="true" tabindex="-1"></a>    <span class="bu">dict</span>[<span class="st">"abs"</span>] <span class="op">=</span> <span class="bu">abs</span></span>
<span id="cb14-62"><a href="#cb14-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-63"><a href="#cb14-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="bu">dict</span></span>
<span id="cb14-64"><a href="#cb14-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-65"><a href="#cb14-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-66"><a href="#cb14-66" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="prepare-input-data" class="level2">
<h2 class="anchored" data-anchor-id="prepare-input-data">Prepare input data</h2>
<p>We want to predict epigenome around HINT2 TSS on chromosome 9.</p>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>targets_txt <span class="op">=</span> <span class="st">'https://raw.githubusercontent.com/calico/basenji/master/manuscripts/cross2020/targets_human.txt'</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>df_targets <span class="op">=</span> pd.read_csv(targets_txt, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>track_names <span class="op">=</span> df_targets[<span class="st">'description'</span>].tolist()</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>track_names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">
<pre><code>['DNASE:cerebellum male adult (27 years) and male adult (35 years)',
 'DNASE:frontal cortex male adult (27 years) and male adult (35 years)',
 'DNASE:chorion',
 'DNASE:Ishikawa treated with 0.02% dimethyl sulfoxide for 1 hour',
 'DNASE:GM03348',
 'DNASE:GM03348 genetically modified using transduction treated with 3 ug/mL doxycycline for 10 days',
 'DNASE:AG08395',
 'DNASE:AG08396',
 'DNASE:AG20443',
 'DNASE:frontal cortex female adult (67 years) and female adult (80 years)',
 'DNASE:H54',
 'DNASE:GM10248',
 'DNASE:GM12878',
 'DNASE:GM12891',
 'DNASE:GM12892',
 'DNASE:GM18507',
 'DNASE:GM19238',
 'DNASE:GM19239',
 'DNASE:GM19240',
 'DNASE:H1-hESC',
 'DNASE:H7-hESC',
 'DNASE:H9',
 'DNASE:heart male adult (27 years) and male adult (35 years)',
 'DNASE:HEK293T',
 'DNASE:HeLa-S3 treated with interferon alpha for 4 hours',
 'DNASE:HeLa-S3',
 'DNASE:hepatocyte',
 'DNASE:HepG2',
 'DNASE:HTR-8/SVneo',
 'DNASE:endothelial cell of umbilical vein newborn',
 'DNASE:CWRU1 male',
 'DNASE:iPS-NIHi11 male adult (71 year) originated from AG20443',
 'DNASE:iPS-NIHi7 female adult (85 years) originated from AG08395',
 'DNASE:K562 treated with 1 uM vorinostat for 72 hours',
 'DNASE:K562 G2 phase',
 'DNASE:K562 G1 phase',
 'DNASE:LNCaP clone FGC',
 'DNASE:LNCaP clone FGC treated with 1 nM 17B-hydroxy-17-methylestra-4,9,11-trien-3-one for 12 hours',
 'DNASE:MCF-7 originated from MCF-7',
 'DNASE:medulloblastoma',
 'DNASE:epidermal melanocyte',
 'DNASE:CD14-positive monocyte female',
 'DNASE:keratinocyte female',
 'DNASE:osteoblast',
 'DNASE:psoas muscle male adult (27 years) and male adult (35 years)',
 'DNASE:T47D treated with 10 nM 17B-estradiol for 30 minutes',
 'DNASE:urothelium cell line',
 'DNASE:A549',
 'DNASE:AG04449',
 'DNASE:AG04450',
 'DNASE:AG09309',
 'DNASE:AG09319',
 'DNASE:AG10803',
 'DNASE:fibroblast of the aortic adventitia female',
 'DNASE:BE2C',
 'DNASE:BJ',
 'DNASE:HS-27A',
 'DNASE:HS-5',
 'DNASE:stromal cell of bone marrow male',
 'DNASE:Caco-2',
 'DNASE:B cell female adult (43 years)',
 'DNASE:hematopoietic multipotent progenitor cell',
 'DNASE:naive thymus-derived CD4-positive, alpha-beta T cell male adult (26 years)',
 'DNASE:CMK',
 'DNASE:GM04503',
 'DNASE:GM04504',
 'DNASE:GM06990',
 'DNASE:GM12864',
 'DNASE:GM12865',
 'DNASE:GM12878',
 'DNASE:H1-hESC',
 'DNASE:cardiac mesoderm',
 'DNASE:cardiac mesoderm',
 'DNASE:cardiac mesoderm',
 'DNASE:cardiac mesoderm',
 'DNASE:H7-hESC',
 'DNASE:astrocyte of the hippocampus',
 'DNASE:astrocyte of the spinal cord',
 'DNASE:astrocyte of the cerebellum',
 'DNASE:amniotic epithelial cell',
 'DNASE:brain microvascular endothelial cell',
 'DNASE:brain pericyte',
 'DNASE:smooth muscle cell of the brain vasculature female',
 'DNASE:cardiac fibroblast',
 'DNASE:cardiac fibroblast female',
 'DNASE:cardiac muscle cell',
 'DNASE:fibroblast of the conjunctiva',
 'DNASE:choroid plexus epithelial cell',
 'DNASE:HCT116',
 'DNASE:epithelial cell of esophagus',
 'DNASE:HeLa-S3 G1b phase',
 'DNASE:HepG2',
 'DNASE:foreskin fibroblast male newborn',
 'DNASE:HFF-Myc originated from foreskin fibroblast',
 'DNASE:fibroblast of gingiva',
 'DNASE:iris pigment epithelial cell',
 'DNASE:HL-60',
 'DNASE:mammary epithelial cell female',
 'DNASE:fibroblast of mammary gland female',
 'DNASE:dermis blood vessel endothelial cell female adult',
 'DNASE:dermis blood vessel endothelial cell female adult',
 'DNASE:dermis blood vessel endothelial cell male newborn',
 'DNASE:dermis microvascular lymphatic vessel endothelial cell female',
 'DNASE:dermis microvascular lymphatic vessel endothelial cell male',
 'DNASE:dermis blood vessel endothelial cell male newborn',
 'DNASE:lung microvascular endothelial cell female',
 'DNASE:lung microvascular endothelial cell female',
 'DNASE:non-pigmented ciliary epithelial cell',
 'DNASE:pulmonary artery endothelial cell female',
 'DNASE:fibroblast of pulmonary artery',
 'DNASE:fibroblast of peridontal ligament male',
 'DNASE:fibroblast of lung',
 'DNASE:renal cortical epithelial cell',
 'DNASE:kidney epithelial cell',
 'DNASE:glomerular endothelial cell',
 'DNASE:retinal pigment epithelial cell',
 'DNASE:skeletal muscle myoblast',
 'DNASE:myotube originated from skeletal muscle myoblast',
 'DNASE:endothelial cell of umbilical vein newborn',
 'DNASE:fibroblast of villous mesenchyme',
 'DNASE:Jurkat clone E61',
 'DNASE:K562',
 'DNASE:K562',
 'DNASE:K562',
 'DNASE:myocyte originated from LHCN-M2',
 'DNASE:LHCN-M2',
 'DNASE:LNCaP clone FGC',
 'DNASE:M059J',
 'DNASE:MCF-7',
 'DNASE:MCF-7 treated with 100 nM estradiol for 1 hour',
 'DNASE:MCF-7',
 'DNASE:CD14-positive monocyte female',
 'DNASE:NB4',
 'DNASE:astrocyte',
 'DNASE:bronchial epithelial cell female treated with retinoic acid',
 'DNASE:fibroblast of dermis female adult',
 'DNASE:foreskin fibroblast male newborn',
 'DNASE:keratinocyte female',
 'DNASE:fibroblast of lung male adult (45 years)',
 'DNASE:NT2/D1',
 'DNASE:Panc1',
 'DNASE:epithelial cell of prostate',
 'DNASE:RPMI-7951',
 'DNASE:epithelial cell of proximal tubule',
 'DNASE:bronchial epithelial cell',
 'DNASE:SK-N-MC',
 'DNASE:SK-N-SH treated with 6 uM all-trans-retinoic acid for 48 hours',
 'DNASE:skeletal muscle cell',
 'DNASE:T47D',
 'DNASE:T-helper 1 cell',
 'DNASE:T-helper 1 cell female adult (26 years)',
 'DNASE:T-helper 1 cell male adult (33 years)',
 'DNASE:T-helper 17 cell',
 'DNASE:T-helper 2 cell',
 'DNASE:T-helper 2 cell female adult (26 years)',
 'DNASE:T-helper 2 cell male adult (33 years)',
 'DNASE:regulatory T cell female adult (35 years)',
 'DNASE:regulatory T cell male adult (28 years)',
 'DNASE:WERI-Rb-1',
 'DNASE:WI38 genetically modified using stable transfection originated from WI38 treated with 20 nM afimoxifene for 72 hours',
 'DNASE:WI38 genetically modified using stable transfection originated from WI38',
 'DNASE:HT1080',
 'DNASE:SK-MEL-5',
 'DNASE:SJCRH30',
 'DNASE:NCI-H460',
 'DNASE:SK-N-DZ treated with dimethyl sulfoxide for 72 hours',
 'DNASE:GM23338 male adult (53 years) originated from GM23248',
 'DNASE:lung embryo (112 days)',
 'DNASE:SK-MEL-5',
 'DNASE:muscle of back female embryo (113 days)',
 'DNASE:iPS DF 4.7 male newborn',
 'DNASE:right renal cortex interstitium male embryo (105 days)',
 'DNASE:caudate nucleus male adult (78 years)',
 'DNASE:left renal pelvis male embryo (105 days)',
 'DNASE:H9 S1 phase genetically modified using stable transfection',
 'DNASE:right kidney female embryo (87 days)',
 'DNASE:Karpas-422',
 'DNASE:CD8-positive, alpha-beta T cell male adult (21 year)',
 'DNASE:renal cortex interstitium female embryo (96 days)',
 'DNASE:brain female embryo (85 days)',
 'DNASE:muscle of leg male embryo (97 days)',
 'DNASE:muscle of back female embryo (105 days)',
 'DNASE:placenta female embryo (113 days)',
 'DNASE:foreskin keratinocyte male newborn',
 'DNASE:right kidney female embryo (107 days)',
 'DNASE:RCC 7860',
 'DNASE:colon epithelial cell line',
 'DNASE:left renal cortex interstitium male embryo (120 days)',
 'DNASE:T-cell',
 'DNASE:large intestine male embryo (113 days)',
 'DNASE:PC-3',
 'DNASE:muscle of leg male embryo (96 days)',
 'DNASE:adrenal gland embryo (96 days)',
 'DNASE:coronary artery female adult (53 years)',
 'DNASE:stomach female embryo (105 days)',
 'DNASE:muscle of arm female embryo (115 days)',
 'DNASE:heart female embryo (91 day)',
 'DNASE:kidney embryo (80 days)',
 'DNASE:heart left ventricle female adult (53 years)',
 'DNASE:muscle of back female embryo (98 days)',
 'DNASE:lung male embryo (108 days)',
 'DNASE:tibial artery female adult (53 years)',
 'DNASE:adrenal gland male embryo (101 day)',
 'DNASE:kidney female embryo (121 day)',
 'DNASE:stomach female embryo (107 days)',
 'DNASE:small intestine female embryo (110 days)',
 'DNASE:MG63',
 'DNASE:skin of body female embryo (82 days)',
 'DNASE:left renal cortex interstitium male embryo (105 days)',
 'DNASE:stomach male embryo (58 days) and male embryo (76 days)',
 'DNASE:left kidney female embryo (59 days) and male embryo (91 day)',
 'DNASE:hematopoietic multipotent progenitor cell',
 'DNASE:kidney female embryo (120 days)',
 'DNASE:muscle of back female embryo (115 days)',
 'DNASE:hematopoietic multipotent progenitor cell male adult (25 years) treated with erythropoietin for 20 days, hydrocortisone succinate for 20 days, kit ligand for 20 days, interleukin-3 for 20 days',
 'DNASE:small intestine female embryo (120 days)',
 'DNASE:brain embryo (112 days)',
 'DNASE:stomach female embryo (105 days)',
 'DNASE:left kidney female embryo (98 days)',
 'DNASE:trophoblast cell embryo (21 week)',
 'DNASE:common myeloid progenitor, CD34-positive female adult (27 years)',
 'DNASE:eye female embryo (76 days)',
 'DNASE:renal cortex interstitium male embryo (108 days)',
 'DNASE:placenta female embryo (108 days)',
 'DNASE:stomach female adult (53 years)',
 'DNASE:small intestine female embryo (107 days)',
 'DNASE:large intestine female embryo (120 days)',
 'DNASE:kidney glomerular epithelial cell male adult (43 years) and male adult (62 years)',
 'DNASE:renal pelvis female embryo (96 days)',
 'DNASE:lung female embryo (76 days)',
 'DNASE:muscle of back male embryo (104 days)',
 'DNASE:lung female embryo (85 days)',
 'DNASE:spinal cord male embryo (105 days)',
 'DNASE:hematopoietic multipotent progenitor cell treated with interleukin-3 for 4 days, kit ligand for 4 days, hydrocortisone succinate for 4 days, erythropoietin for 4 days',
 'DNASE:HepG2',
 'DNASE:foreskin fibroblast male newborn',
 'DNASE:omental fat pad female adult (53 years)',
 'DNASE:large intestine female embryo (110 days)',
 'DNASE:heart female embryo (105 days)',
 'DNASE:renal cortex interstitium male embryo (113 days)',
 'DNASE:brain female embryo (96 days)',
 'DNASE:thyroid gland male adult (37 years)',
 'DNASE:liver embryo (59 days) and embryo (80 days)',
 'DNASE:stomach male adult (54 years)',
 'DNASE:thymus male embryo (127 days)',
 'DNASE:upper lobe of left lung male adult (37 years)',
 'DNASE:endodermal cell',
 'DNASE:forelimb muscle female embryo (108 days)',
 'DNASE:CD4-positive, alpha-beta T cell male adult (37 years)',
 'DNASE:muscle of leg male embryo (101 day)',
 'DNASE:gastrocnemius medialis male adult (37 years)',
 'DNASE:gastrocnemius medialis male adult (54 years)',
 'DNASE:stomach male embryo (91 day)',
 'DNASE:muscle of back male embryo (101 day)',
 'DNASE:heart female embryo (117 days)',
 'DNASE:kidney tubule cell female adult (80 years) and male adult (62 years)',
 'DNASE:fibroblast of skin of abdomen male embryo (97 days)',
 'DNASE:pancreas male adult (34 years)',
 'DNASE:trophoblast cell originated from H1-hESC',
 'DNASE:left kidney male embryo (96 days)',
 'DNASE:skin fibroblast male embryo (97 days)',
 'DNASE:brain female embryo (142 days)',
 'DNASE:small intestine male embryo (105 days)',
 'DNASE:heart female embryo (110 days)',
 'DNASE:common myeloid progenitor, CD34-positive female',
 'DNASE:adrenal gland male adult (37 years)',
 'DNASE:muscle of back female embryo (85 days)',
 'DNASE:muscle of leg female embryo (85 days)',
 "DNASE:Peyer's patch male adult (54 years)",
 'DNASE:amniotic stem cell',
 'DNASE:stomach female embryo (98 days)',
 'DNASE:large intestine female embryo (91 day)',
 'DNASE:lung female embryo (120 days)',
 'DNASE:skin fibroblast male embryo (97 days)',
 'DNASE:SW480',
 'DNASE:GM23248',
 'DNASE:muscle of trunk female embryo (120 days)',
 'DNASE:islet precursor cell',
 'DNASE:medulla oblongata male adult (78 years) and male adult (84 years)',
 'DNASE:right lung male embryo (115 days)',
 'DNASE:body of pancreas male adult (54 years)',
 'DNASE:mammary epithelial cell female adult (18 years)',
 'DNASE:gastrocnemius medialis female adult (51 year)',
 'DNASE:uterus female adult (53 years)',
 'DNASE:ELR',
 'DNASE:omental fat pad female adult (51 year)',
 'DNASE:stomach male child (3 years)',
 'DNASE:B cell male adult (37 years)',
 'DNASE:foreskin fibroblast male newborn',
 'DNASE:renal cortex interstitium male embryo (91 day)',
 'DNASE:ACHN',
 'DNASE:kidney tubule cell female adult (80 years) treated with 5 uM cisplatin',
 'DNASE:left renal pelvis male embryo (105 days)',
 'DNASE:iPS DF 6.9 male newborn',
 'DNASE:kidney male embryo (87 days)',
 'DNASE:right lung female embryo (91 day)',
 'DNASE:NCI-H226',
 'DNASE:right renal cortex interstitium male embryo (105 days)',
 'DNASE:G401',
 'DNASE:muscle of arm male embryo (113 days)',
 'DNASE:common myeloid progenitor, CD34-positive male adult',
 'DNASE:muscle of leg female embryo (113 days)',
 'DNASE:arm bone male embryo (81 day)',
 'DNASE:H9',
 'DNASE:testis male embryo',
 'DNASE:neural stem progenitor cell originated from H1-hESC',
 'DNASE:right atrium auricular region female adult (51 year)',
 'DNASE:common myeloid progenitor, CD34-positive male adult (49 years)',
 'DNASE:stomach female embryo (121 day)',
 'DNASE:muscle of leg female embryo (105 days)',
 'DNASE:muscle of leg male embryo (97 days)',
 'DNASE:left kidney female embryo (87 days)',
 'DNASE:muscle of back male embryo (97 days)',
 'DNASE:muscle of leg male embryo (96 days)',
 'DNASE:right lung female embryo (105 days)',
 'DNASE:NAMALWA treated with Sendai virus for 2 hours',
 'DNASE:renal pelvis male embryo (127 days)',
 'DNASE:muscle of arm male embryo (96 days)',
 'DNASE:heart male child (3 years)',
 'DNASE:brain female embryo (117 days)',
 'DNASE:muscle of arm female embryo (85 days)',
 'DNASE:heart male embryo (110 days)',
 'DNASE:L1-S8',
 'DNASE:CD8-positive, alpha-beta T cell male adult (37 years)',
 'DNASE:muscle of back male embryo (91 day)',
 'DNASE:left renal pelvis male embryo (105 days)',
 'DNASE:middle frontal gyrus male adult (78 years)',
 'DNASE:lung male embryo (103 days)',
 'DNASE:heart male embryo (105 days)',
 'DNASE:left renal pelvis male embryo (120 days)',
 'DNASE:common myeloid progenitor, CD34-positive female adult (50 years)',
 "DNASE:Peyer's patch female adult (53 years)",
 'DNASE:skin fibroblast male embryo (97 days)',
 'DNASE:hematopoietic multipotent progenitor cell',
 'DNASE:right kidney male embryo (87 days)',
 'DNASE:tibial artery male adult (37 years)',
 'DNASE:renal cell carcinoma',
 'DNASE:thymus female embryo (113 days)',
 'DNASE:brain female embryo (109 days)',
 'DNASE:Daoy',
 'DNASE:A673',
 'DNASE:muscle of arm male embryo (101 day)',
 'DNASE:muscle of arm female embryo (120 days)',
 'DNASE:muscle of back male embryo (105 days)',
 'DNASE:skin fibroblast male embryo (97 days)',
 'DNASE:ecto neural progenitor cell originated from H9',
 'DNASE:hematopoietic multipotent progenitor cell treated with interleukin-3 for 17 days, kit ligand for 17 days, hydrocortisone succinate for 17 days, erythropoietin for 17 days',
 'DNASE:hepatocyte originated from H9',
 'DNASE:renal pelvis male embryo (97 days)',
 'DNASE:heart embryo (101 day)',
 'DNASE:large intestine male embryo (115 days)',
 'DNASE:T-cell male adult (37 years)',
 'DNASE:skin fibroblast male embryo (97 days)',
 'DNASE:CD1c-positive myeloid dendritic cell',
 'DNASE:B cell male adult (21 year)',
 'DNASE:superior temporal gyrus male adult (84 years)',
 'DNASE:skin fibroblast male embryo (97 days)',
 'DNASE:iPS DF 19.11 male newborn',
 'DNASE:hindlimb muscle male embryo (120 days)',
 'DNASE:MCF 10A treated with 1 uM tamoxifen for 6 hours',
 'DNASE:sigmoid colon male adult (54 years)',
 'DNASE:heart male embryo (120 days)',
 'DNASE:tibial nerve female adult (51 year)',
 'DNASE:adipocyte',
 'DNASE:muscle of trunk female embryo (121 day)',
 'DNASE:left cardiac atrium female embryo (101 day)',
 'DNASE:T-cell male adult (21 year)',
 'DNASE:right kidney male embryo (108 days)',
 'DNASE:adrenal gland female embryo (108 days)',
 'DNASE:hematopoietic multipotent progenitor cell treated with interleukin-3 for 11 day, kit ligand for 11 day, hydrocortisone succinate for 11 day, erythropoietin for 11 day',
 'DNASE:brain male embryo (105 days)',
 'DNASE:body of pancreas female adult (53 years)',
 'DNASE:left kidney female embryo (147 days)',
 'DNASE:CD4-positive, alpha-beta T cell male adult (37 years)',
 'DNASE:KBM-7',
 'DNASE:psoas muscle male child (3 years)',
 'DNASE:ovary female embryo',
 'DNASE:RKO',
 'DNASE:leg bone male embryo (81 day)',
 'DNASE:glomerular visceral epithelial cell child (3 years)',
 'DNASE:foreskin melanocyte male newborn',
 'DNASE:embryonic facial prominence embryo (53 days) and embryo (58 days)',
 'DNASE:vagina female adult (53 years)',
 'DNASE:kidney female embryo (85 days)',
 'DNASE:muscle of back male embryo (96 days)',
 'DNASE:muscle of trunk female embryo (113 days)',
 'DNASE:iPS DF 19.7 male newborn',
 'DNASE:right lung female embryo (110 days)',
 'DNASE:left lung male embryo (87 days)',
 'DNASE:large intestine female embryo (105 days)',
 'DNASE:right lung female embryo (98 days)',
 'DNASE:leg bone male embryo (81 day)',
 'DNASE:CD14-positive monocyte male adult (21 year)',
 'DNASE:muscle of arm male embryo (120 days)',
 "DNASE:Peyer's patch male adult (37 years)",
 'DNASE:common myeloid progenitor, CD34-positive male adult (36 years)',
 'DNASE:left lung female embryo (108 days)',
 'DNASE:small intestine female embryo (91 day)',
 'DNASE:MM.1S',
 'DNASE:sigmoid colon female adult (51 year)',
 'DNASE:left lung female embryo (105 days)',
 'DNASE:common myeloid progenitor, CD34-positive female adult (33 years)',
 'DNASE:retina embryo (125 days) and male embryo (103 days)',
 'DNASE:brain male embryo (72 days) and male embryo (76 days)',
 'DNASE:muscle of back male embryo (108 days)',
 'DNASE:IMR-90',
 'DNASE:lung female embryo (108 days)',
 'DNASE:left lung male embryo (113 days)',
 'DNASE:OCI-LY7',
 'DNASE:upper lobe of left lung female adult (51 year)',
 'DNASE:hematopoietic multipotent progenitor cell treated with interleukin-3 for 18 days, kit ligand for 18 days, hydrocortisone succinate for 18 days, erythropoietin for 18 days',
 'DNASE:putamen male adult (78 years)',
 'DNASE:common myeloid progenitor, CD34-positive male adult (42 years)',
 'DNASE:placenta embryo (53 days)',
 'DNASE:kidney male embryo (105 days)',
 'DNASE:heart left ventricle female embryo (136 days)',
 'DNASE:small intestine female embryo (108 days)',
 'DNASE:LoVo',
 'DNASE:cerebellar cortex male adult (78 years) and male adult (84 years)',
 'DNASE:lung embryo (67 days)',
 'DNASE:transverse colon female adult (53 years)',
 'DNASE:brain embryo (80 days)',
 'DNASE:stomach female embryo (108 days)',
 'DNASE:umbilical cord embryo (59 days) and male embryo (76 days)',
 'DNASE:muscle of leg male embryo (105 days)',
 'DNASE:Caki2',
 'DNASE:muscle of arm female embryo (105 days)',
 'DNASE:foreskin melanocyte male newborn',
 'DNASE:stomach male embryo (108 days)',
 'DNASE:body of pancreas male adult (37 years)',
 'DNASE:renal cortex interstitium female embryo (120 days)',
 'DNASE:body of pancreas female adult (51 year)',
 'DNASE:large intestine male embryo (105 days)',
 'DNASE:limb embryo (58 days) and embryo (59 days)',
 'DNASE:heart female embryo (147 days)',
 'DNASE:heart male embryo (96 days)',
 'DNASE:right renal pelvis male embryo (120 days)',
 'DNASE:left lung female embryo (117 days)',
 'DNASE:thymus female embryo (98 days)',
 'DNASE:ovary female adult (53 years)',
 'DNASE:left kidney male embryo (115 days)',
 'DNASE:thymus male embryo (113 days)',
 'DNASE:placenta female embryo (105 days)',
 'DNASE:EH',
 'DNASE:renal cortex interstitium male embryo (97 days)',
 'DNASE:placenta embryo (102 days)',
 'DNASE:placenta embryo (56 days) and embryo (59 days)',
 'DNASE:renal pelvis female embryo (103 days)',
 'DNASE:right lobe of liver female adult (53 years)',
 'DNASE:skin fibroblast male embryo (97 days)',
 'DNASE:tongue male embryo (72 days)',
 'DNASE:skin fibroblast male embryo (97 days)',
 'DNASE:liver female embryo (101 day) and female embryo (113 days)',
 'DNASE:large intestine female embryo (107 days)',
 'DNASE:hematopoietic multipotent progenitor cell treated with interleukin-3 for 6 days, kit ligand for 6 days, hydrocortisone succinate for 6 days, erythropoietin for 6 days',
 'DNASE:common myeloid progenitor, CD34-positive male',
 'DNASE:small intestine female embryo (98 days)',
 'DNASE:CD4-positive, alpha-beta T cell male adult (21 year)',
 'DNASE:brain male embryo (101 day)',
 'DNASE:placenta female embryo (101 day) and male embryo (105 days)',
 'DNASE:renal pelvis male embryo (91 day)',
 'DNASE:lung embryo (80 days) and male embryo (76 days)',
 "DNASE:Ammon's horn male adult (84 years)",
 'DNASE:muscle of back female embryo (105 days)',
 'DNASE:lung male embryo (82 days)',
 'DNASE:left renal cortex interstitium male embryo (105 days)',
 'DNASE:trophoblast cell embryo (17 weeks) and embryo (18 weeks)',
 'DNASE:RPMI8226',
 'DNASE:small intestine male embryo (91 day)',
 'DNASE:brain embryo (56 days) and male embryo (58 days)',
 'DNASE:pancreas female adult (30 years)',
 'DNASE:adrenal gland female adult (53 years)',
 'DNASE:breast epithelium female adult (51 year)',
 'DNASE:left kidney male embryo (87 days)',
 'DNASE:renal pelvis female embryo (96 days)',
 'DNASE:stomach male embryo (127 days)',
 'DNASE:CD8-positive, alpha-beta T cell female adult (34 years)',
 'DNASE:spinal cord female embryo (89 days)',
 'DNASE:muscle of arm male embryo (97 days)',
 'DNASE:renal pelvis female embryo (105 days)',
 'DNASE:HAP-1',
 'DNASE:SJSA1',
 'DNASE:retina embryo (74 days) and embryo (85 days)',
 'DNASE:muscle of leg male embryo (127 days)',
 'DNASE:fibroblast of skin of abdomen male embryo (97 days)',
 'DNASE:bipolar neuron originated from GM23338 treated with 0.5 ug/mL doxycycline hyclate for 4 days',
 'DNASE:stomach female embryo (147 days)',
 'DNASE:lung male embryo (54 days) and male embryo (58 days)',
 'DNASE:T-cell male adult (36 years)',
 'DNASE:T-cell male adult (21 year)',
 'DNASE:HK-2',
 'DNASE:NCI-H460',
 'DNASE:left lung male embryo (96 days)',
 'DNASE:common myeloid progenitor, CD34-positive male adult (23 years)',
 'DNASE:stomach female adult (51 year)',
 'DNASE:CD4-positive, alpha-beta T cell female adult (33 years)',
 'DNASE:kidney female embryo (113 days)',
 'DNASE:skin fibroblast male embryo (97 days)',
 'DNASE:placenta female embryo (85 days)',
 'DNASE:brain male embryo (122 days)',
 'DNASE:muscle of arm male embryo (97 days)',
 'DNASE:renal cortex interstitium female embryo (103 days)',
 'DNASE:thyroid gland female adult (51 year)',
 'DNASE:NAMALWA',
 'DNASE:heart embryo (59 days) and female embryo (76 days)',
 'DNASE:large intestine male embryo (105 days)',
 'DNASE:skin fibroblast male embryo (97 days)',
 'DNASE:right kidney male embryo (96 days)',
 'DNASE:CD14-positive monocyte male adult (37 years)',
 'DNASE:ELF-1',
 'DNASE:spinal cord female embryo (113 days)',
 'DNASE:spleen embryo (112 days)',
 'DNASE:left lung female embryo (107 days)',
 'DNASE:thymus female embryo',
 'DNASE:CD4-positive, alpha-beta T cell male adult (21 year)',
 'DNASE:stomach female embryo (96 days)',
 'DNASE:EL',
 'DNASE:CD14-positive monocyte female adult (34 years)',
 'DNASE:thyroid gland female adult (53 years)',
 'DNASE:adrenal gland female embryo (113 days)',
 'DNASE:muscle of back male embryo (96 days)',
 'DNASE:SJCRH30',
 'DNASE:natural killer cell male adult (21 year)',
 'DNASE:heart male embryo (72 days) and male embryo (76 days)',
 'DNASE:midbrain male adult (78 years) and male adult (84 years)',
 'DNASE:ovary female adult (30 years)',
 'DNASE:mesendoderm originated from H1-hESC',
 'DNASE:natural killer cell male adult (37 years)',
 'DNASE:natural killer cell female adult (34 years)',
 'DNASE:foreskin keratinocyte male newborn',
 'DNASE:stomach female embryo',
 'DNASE:adrenal gland male embryo (108 days)',
 'DNASE:testis male embryo',
 'DNASE:mesenchymal stem cell originated from H1-hESC',
 'DNASE:vagina female adult (51 year)',
 'DNASE:common myeloid progenitor, CD34-positive male adult (43 years)',
 'DNASE:heart right ventricle female embryo (101 day) and female embryo (103 days)',
 'DNASE:L1-S8R',
 'DNASE:heart female embryo (116 days) and female embryo (98 days)',
 'DNASE:MCF 10A treated with 1 uM tamoxifen for 24 hours',
 'DNASE:right renal pelvis male embryo (105 days)',
 'DNASE:muscle of arm female embryo (98 days)',
 'DNASE:H9 G1 phase genetically modified using stable transfection',
 'DNASE:tibial nerve male adult (37 years)',
 'DNASE:kidney male embryo (85 days)',
 'DNASE:tongue female embryo (59 days) and female embryo (76 days)',
 'DNASE:transverse colon female adult (51 year)',
 'DNASE:right renal pelvis male embryo (105 days)',
 'DNASE:large intestine male embryo (91 day)',
 'DNASE:right kidney female embryo (117 days)',
 'DNASE:right kidney male embryo (91 day)',
 'DNASE:skin fibroblast male embryo (97 days)',
 'DNASE:globus pallidus male adult (78 years) and male adult (84 years)',
 'DNASE:kidney female embryo (105 days)',
 'DNASE:right atrium auricular region female adult (53 years)',
 'DNASE:thymus female embryo (147 days)',
 'DNASE:stomach male adult (34 years)',
 'DNASE:eye embryo (56 days) and male embryo (76 days)',
 'DNASE:esophagus squamous epithelium male adult (37 years)',
 'DNASE:glomerular visceral epithelial cell child (3 years)',
 'DNASE:muscle of back male embryo (127 days)',
 'DNASE:spinal cord male embryo (96 days)',
 'DNASE:MCF 10A',
 'DNASE:transverse colon male adult (54 years)',
 'DNASE:kidney embryo (59 days) and female embryo (59 days)',
 'DNASE:H1-hESC',
 'DNASE:large intestine female embryo (98 days)',
 'DNASE:large intestine female embryo (103 days)',
 'DNASE:placenta male embryo (85 days)',
 'DNASE:left lung female embryo (91 day)',
 'DNASE:heart embryo (80 days)',
 'DNASE:femur female embryo (98 days)',
 'DNASE:H4',
 'DNASE:trophoblast cell embryo (23 weeks)',
 'DNASE:urinary bladder male embryo (76 days)',
 'DNASE:HT-29',
 'DNASE:limb embryo (53 days) and embryo (56 days)',
 'DNASE:retina female embryo (89 days)',
 'DNASE:dedifferentiated amniotic fluid mesenchymal stem cell',
 'DNASE:upper lobe of left lung female adult (53 years)',
 'DNASE:brain female embryo (105 days)',
 'DNASE:muscle of arm embryo (101 day)',
 'DNASE:renal cortex interstitium male embryo (108 days)',
 'DNASE:pons male adult (78 years)',
 'DNASE:thymus female embryo (105 days)',
 'DNASE:CD8-positive, alpha-beta T cell female adult (33 years)',
 'DNASE:cardiac muscle cell',
 'DNASE:CD8-positive, alpha-beta T cell male adult (21 year)',
 'DNASE:hematopoietic multipotent progenitor cell treated with interleukin-3 for 8 days, kit ligand for 8 days, hydrocortisone succinate for 8 days, erythropoietin for 8 days',
 'DNASE:right lung female embryo (117 days)',
 'DNASE:right lung male embryo (105 days)',
 'DNASE:lung female embryo (96 days)',
 'DNASE:adrenal gland male adult (54 years)',
 'DNASE:renal cortex interstitium male embryo (127 days)',
 'DNASE:spleen male adult (54 years)',
 'DNASE:renal cortex interstitium female embryo (89 days)',
 'DNASE:ascending aorta female adult (51 year)',
 'DNASE:ovary female adult (51 year)',
 'DNASE:hematopoietic multipotent progenitor cell treated with interleukin-3 for 13 days, kit ligand for 13 days, hydrocortisone succinate for 13 days, erythropoietin for 13 days',
 'DNASE:cardiac fibroblast female embryo (94 days) and female embryo (98 days)',
 'DNASE:adrenal gland male embryo (85 days)',
 'DNASE:large intestine male embryo (108 days)',
 'DNASE:occipital lobe male adult (84 years)',
 'DNASE:right lung female embryo (107 days)',
 'DNASE:small intestine female embryo (105 days)',
 'DNASE:muscle of leg male embryo (104 days)',
 'DNASE:lower leg skin female adult (53 years)',
 'DNASE:adrenal gland female adult (51 year)',
 'DNASE:adrenal gland female embryo (85 days)',
 'DNASE:muscle of arm male embryo (115 days)',
 'DNASE:right kidney female embryo (147 days)',
 'DNASE:kidney female embryo (76 days) and male embryo (76 days)',
 'DNASE:large intestine female embryo (108 days)',
 'DNASE:right kidney female embryo (98 days)',
 'DNASE:right renal cortex interstitium male embryo (120 days)',
 'DNASE:muscle of arm male embryo (96 days)',
 'DNASE:thymus male embryo (108 days)',
 'DNASE:B cell female adult (34 years)',
 'DNASE:thyroid gland male adult (54 years)',
 'DNASE:spinal cord female embryo (59 days) and male embryo (72 days)',
 'DNASE:heart embryo (96 days)',
 'DNASE:renal pelvis female embryo (89 days)',
 'DNASE:renal pelvis male embryo (108 days)',
 'DNASE:heart female embryo (103 days)',
 'DNASE:H9',
 'DNASE:K562',
 'DNASE:left renal cortex interstitium male embryo (105 days)',
 'DNASE:right lung female embryo (108 days)',
 'DNASE:stomach female embryo (96 days)',
 'DNASE:small intestine male adult (34 years)',
 'DNASE:renal cortex interstitium female embryo (96 days)',
 'DNASE:A172',
 'DNASE:left lung male embryo (115 days)',
 'DNASE:muscle of leg male embryo (115 days)',
 'DNASE:hematopoietic multipotent progenitor cell treated with interleukin-3 for 15 days, kit ligand for 15 days, hydrocortisone succinate for 15 days, erythropoietin for 15 days',
 'DNASE:muscle of leg female embryo (115 days)',
 'DNASE:PC-9',
 'DNASE:kidney female embryo (108 days)',
 'DNASE:stomach embryo (101 day)',
 'DNASE:H9 G2 phase genetically modified using stable transfection',
 'DNASE:esophagus muscularis mucosa male adult (37 years)',
 'DNASE:small intestine male embryo (108 days)',
 'DNASE:heart left ventricle female embryo (101 day) and female embryo (103 days)',
 'DNASE:SK-N-DZ',
 'DNASE:brain male embryo (104 days)',
 'DNASE:right lung male embryo (87 days)',
 'DNASE:muscle of arm male embryo (105 days)',
 'DNASE:muscle of arm male embryo (104 days)',
 'DNASE:kidney female embryo (105 days)',
 'DNASE:left lung male embryo (105 days)',
 'DNASE:left kidney female embryo (107 days)',
 "DNASE:Peyer's patch male adult (37 years)",
 'DNASE:renal pelvis male embryo (113 days)',
 'DNASE:prostate gland male adult (37 years)',
 'DNASE:HeLa-S3',
 'DNASE:left lung female embryo (110 days)',
 'DNASE:thymus male embryo (104 days)',
 'DNASE:neural progenitor cell originated from H9',
 'DNASE:trophoblast cell embryo (39 weeks) and embryo (40 weeks)',
 'DNASE:right lung male embryo (96 days)',
 'DNASE:H9 G1 phase genetically modified using stable transfection',
 'DNASE:small intestine male embryo (87 days)',
 'DNASE:spinal cord female embryo (87 days)',
 'DNASE:left kidney female embryo (110 days)',
 'DNASE:arm bone male embryo (81 day)',
 'DNASE:testis male adult (54 years)',
 'DNASE:transverse colon male adult (37 years)',
 'DNASE:right renal pelvis male embryo (105 days)',
 'DNASE:lung embryo (101 day)',
 'DNASE:inferior parietal cortex male adult (84 years)',
 'DNASE:kidney capillary endothelial cell female embryo (113 days)',
 'DNASE:placenta male embryo (91 day)',
 'DNASE:right kidney male embryo (115 days)',
 'DNASE:coronary artery female adult (51 year)',
 'ATAC:BM0106-Day0-MCP-A / Bone Marrow CD34+ / pDC',
 'ATAC:BM0106-UNK-ATAC-2 / Bone Marrow CD34+ / UNK',
 'ATAC:BM0828-MEGA1-A-151109 / Bone Marrow CD34+ / Mega',
 'ATAC:BM1077-MCP / Bone Marrow CD34+ / pDC',
 'ATAC:BM1077-UNK / Bone Marrow CD34+ / UNK',
 'ATAC:BM1137-GMP1-low-ATAC-2 / Bone Marrow CD34+ / GMP-A',
 'ATAC:BM1137-GMP2-mid-ATAC-1 / Bone Marrow CD34+ / GMP-B',
 'ATAC:BM1137-GMP3-high-ATAC-2 / Bone Marrow CD34+ / GMP-C',
 'ATAC:BM1214-Day0-MCP / Bone Marrow CD34+ / pDC',
 'ATAC:BM1214-Day0-UNK-A / Bone Marrow CD34+ / UNK',
 'CHIP:CTCF:MCF-7',
 'CHIP:TAF1:MCF-7',
 'CHIP:H3K4me3:GM12878',
 'CHIP:CTCF:GM12878',
 'CHIP:H3K27ac:GM12878',
 'CHIP:H3K27me3:GM12878',
 'CHIP:H3K4me1:GM12878',
 'CHIP:H3K4me2:GM12878',
 'CHIP:H3K9ac:GM12878',
 'CHIP:H4K20me1:GM12878',
 'CHIP:H3K27me3:endothelial cell of umbilical vein male newborn',
 'CHIP:H3K4me2:endothelial cell of umbilical vein male newborn',
 'CHIP:H3K4me3:endothelial cell of umbilical vein male newborn',
 'CHIP:CTCF:K562',
 'CHIP:H3K27me3:K562',
 'CHIP:H3K36me3:K562',
 'CHIP:H3K4me2:K562',
 'CHIP:H3K4me3:K562',
 'CHIP:H3K9ac:K562',
 'CHIP:H3K9me1:K562',
 'CHIP:CTCF:endothelial cell of umbilical vein male newborn',
 'CHIP:H3K27ac:endothelial cell of umbilical vein male newborn',
 'CHIP:H3K27ac:keratinocyte female',
 'CHIP:H3K27ac:mammary epithelial cell female adult (50 years)',
 'CHIP:H3K27me3:mammary epithelial cell female adult (50 years)',
 'CHIP:H3K36me3:mammary epithelial cell female adult (50 years)',
 'CHIP:H3K4me2:HepG2',
 'CHIP:H3K9ac:HepG2',
 'CHIP:CTCF:H1-hESC',
 'CHIP:H3K4me1:mammary epithelial cell female adult (50 years)',
 'CHIP:H3K4me2:mammary epithelial cell female adult (50 years)',
 'CHIP:H3K4me3:mammary epithelial cell female adult (50 years)',
 'CHIP:H4K20me1:mammary epithelial cell female adult (50 years)',
 'CHIP:H3K4me3:HepG2',
 'CHIP:H4K20me1:HepG2',
 'CHIP:H3K27ac:fibroblast of lung female child (11 year) and male adult (45 years)',
 'CHIP:H3K27me3:fibroblast of lung female child (11 year) and male adult (45 years)',
 'CHIP:H3K36me3:fibroblast of lung female child (11 year) and male adult (45 years)',
 'CHIP:H3K4me3:fibroblast of lung female child (11 year) and male adult (45 years)',
 'CHIP:H3K9ac:fibroblast of lung female child (11 year) and male adult (45 years)',
 'CHIP:H3K27ac:skeletal muscle myoblast male adult (22 years)',
 'CHIP:H3K27me3:skeletal muscle myoblast male adult (22 years)',
 'CHIP:H3K36me3:skeletal muscle myoblast male adult (22 years)',
 'CHIP:H3K4me1:skeletal muscle myoblast male adult (22 years)',
 'CHIP:H3K4me2:skeletal muscle myoblast male adult (22 years)',
 'CHIP:H3K4me3:skeletal muscle myoblast male adult (22 years)',
 'CHIP:H3K9ac:skeletal muscle myoblast male adult (22 years)',
 'CHIP:H4K20me1:skeletal muscle myoblast male adult (22 years)',
 'CHIP:H3K27ac:H1-hESC',
 'CHIP:H3K79me2:skeletal muscle myoblast male adult (22 years)',
 'CHIP:H3K9me3:skeletal muscle myoblast male adult (22 years)',
 'CHIP:CTCF:myotube',
 'CHIP:H2AFZ:myotube',
 'CHIP:H3K27ac:myotube',
 'CHIP:H3K36me3:myotube',
 'CHIP:H3K4me1:myotube',
 'CHIP:H3K4me2:myotube',
 'CHIP:H3K4me3:myotube',
 'CHIP:CTCF:HeLa-S3',
 'CHIP:H3K27ac:HeLa-S3',
 'CHIP:H3K4me2:HeLa-S3',
 'CHIP:H3K4me3:HeLa-S3',
 'CHIP:H3K79me2:HeLa-S3',
 'CHIP:H3K9ac:HeLa-S3',
 'CHIP:H4K20me1:HeLa-S3',
 'CHIP:H3K27me3:HepG2',
 'CHIP:H3K79me2:myotube',
 'CHIP:CTCF:astrocyte',
 'CHIP:H3K27ac:astrocyte',
 'CHIP:H3K4me3:astrocyte',
 'CHIP:H2AFZ:GM12878',
 'CHIP:H3K79me2:GM12878',
 'CHIP:H3K9ac:myotube',
 'CHIP:H4K20me1:myotube',
 'CHIP:H2AFZ:skeletal muscle myoblast male adult (22 years)',
 'CHIP:H3K27me3:HeLa-S3',
 'CHIP:H2AFZ:K562',
 'CHIP:H3K79me2:K562',
 'CHIP:H3K9me3:K562',
 'CHIP:H2AFZ:osteoblast',
 'CHIP:H3K27ac:osteoblast',
 'CHIP:H3K36me3:osteoblast',
 'CHIP:H3K4me1:osteoblast',
 'CHIP:H3K4me2:osteoblast',
 'CHIP:H3K9me3:osteoblast',
 'CHIP:CTCF:fibroblast of dermis',
 'CHIP:H3K27ac:fibroblast of dermis',
 'CHIP:H3K4me3:fibroblast of dermis NONE and female adult',
 'CHIP:H3K4me1:HepG2',
 'CHIP:H3K4me1:HeLa-S3',
 'CHIP:KDM5B:K562',
 'CHIP:RBBP5:H1-hESC',
 'CHIP:HDAC1:K562',
 'CHIP:HDAC2:K562',
 'CHIP:PHF8:K562',
 'CHIP:RBBP5:K562',
 'CHIP:SAP30:K562',
 'CHIP:CHD1:H1-hESC',
 'CHIP:H3K9me3:myotube',
 'CHIP:H2AFZ:HeLa-S3',
 'CHIP:H3K9me3:HeLa-S3',
 'CHIP:H3K27me3:osteoblast',
 'CHIP:H4K20me1:osteoblast',
 'CHIP:H2AFZ:DND-41',
 'CHIP:H3K27ac:DND-41',
 'CHIP:H3K36me3:DND-41',
 'CHIP:H3K4me1:DND-41',
 'CHIP:H3K4me3:DND-41',
 'CHIP:H3K79me2:DND-41',
 'CHIP:H4K20me1:DND-41',
 'CHIP:EZH2:GM12878',
 'CHIP:EZH2:mammary epithelial cell female adult (50 years)',
 'CHIP:H2AFZ:mammary epithelial cell female adult (50 years)',
 'CHIP:H3K9me3:mammary epithelial cell female adult (50 years)',
 'CHIP:EZH2:skeletal muscle myoblast male adult (22 years)',
 'CHIP:EZH2:HepG2',
 'CHIP:H2AFZ:keratinocyte female',
 'CHIP:H3K79me2:keratinocyte female',
 'CHIP:H3K9me3:keratinocyte female',
 'CHIP:EZH2:fibroblast of lung female child (11 year) and male adult (45 years)',
 'CHIP:H2AFZ:fibroblast of lung female child (11 year) and male adult (45 years)',
 'CHIP:H3K9me3:fibroblast of lung female child (11 year) and male adult (45 years)',
 'CHIP:H3K9ac:astrocyte',
 'CHIP:H4K20me1:astrocyte',
 'CHIP:H3K9ac:DND-41',
 'CHIP:H3K9me3:DND-41',
 'CHIP:H3K79me2:mammary epithelial cell female adult (50 years)',
 'CHIP:H2AFZ:endothelial cell of umbilical vein male newborn',
 'CHIP:H3K79me2:endothelial cell of umbilical vein male newborn',
 'CHIP:EZH2:fibroblast of dermis',
 'CHIP:H3K79me2:fibroblast of lung female child (11 year) and male adult (45 years)',
 'CHIP:H3K79me2:osteoblast',
 'CHIP:H3K4me3:A549 treated with 0.02% ethanol for 1 hour',
 'CHIP:H3K27ac:CD14-positive monocyte female',
 'CHIP:H3K27me3:CD14-positive monocyte female',
 'CHIP:H3K36me3:CD14-positive monocyte female',
 'CHIP:H3K4me1:CD14-positive monocyte female',
 'CHIP:H3K4me3:CD14-positive monocyte female',
 'CHIP:H3K79me2:CD14-positive monocyte female',
 'CHIP:H3K9me3:CD14-positive monocyte female',
 'CHIP:H4K20me1:CD14-positive monocyte female',
 'CHIP:H3K4me3:A549 treated with 100 nM dexamethasone for 1 hour',
 'CHIP:H3K9ac:A549 treated with 0.02% ethanol for 1 hour',
 'CHIP:H3K27me3:DND-41',
 'CHIP:H3K9me3:endothelial cell of umbilical vein male newborn',
 'CHIP:EZH2:HeLa-S3',
 'CHIP:H3K9me3:HepG2',
 'CHIP:H3K9ac:CD14-positive monocyte female',
 'CHIP:H3K4me3:osteoblast',
 'CHIP:H3K27me3:myotube',
 'CHIP:HDAC6:K562',
 'CHIP:PHF8:H1-hESC',
 'CHIP:CTCF:CD14-positive monocyte female',
 'CHIP:H3K79me2:A549 treated with 0.02% ethanol for 1 hour',
 'CHIP:HDAC6:H1-hESC',
 'CHIP:SAP30:H1-hESC',
 'CHIP:SUZ12:H1-hESC',
 'CHIP:CREBBP:K562',
 'CHIP:CBX3:K562',
 'CHIP:KAT2B:K562',
 'CHIP:SIRT6:K562',
 'CHIP:SUZ12:K562',
 'CHIP:CTCF:A549 treated with 100 nM dexamethasone for 1 hour',
 'CHIP:H2AFZ:A549 treated with 0.02% ethanol for 1 hour',
 'CHIP:H3K27ac:A549 treated with 0.02% ethanol for 1 hour',
 'CHIP:H3K27me3:A549 treated with 0.02% ethanol for 1 hour',
 'CHIP:H3K4me1:A549 treated with 0.02% ethanol for 1 hour',
 'CHIP:H3K9me3:A549 treated with 0.02% ethanol for 1 hour',
 'CHIP:H4K20me1:A549 treated with 0.02% ethanol for 1 hour',
 'CHIP:H3K27ac:B cell female adult (27 years)',
 'CHIP:EP300:H1-hESC',
 'CHIP:SIRT6:H1-hESC',
 'CHIP:SETDB1:K562',
 'CHIP:CTCF:B cell female adult (27 years) and female adult (43 years)',
 'CHIP:H2AFZ:B cell female adult (27 years) and female adult (43 years)',
 'CHIP:H3K4me2:B cell female adult (27 years) and female adult (43 years)',
 'CHIP:CHD7:H1-hESC',
 'CHIP:HDAC2:H1-hESC',
 'CHIP:KDM4A:H1-hESC',
 'CHIP:WHSC1:K562',
 'CHIP:H3K27ac:A549 treated with 100 nM dexamethasone for 1 hour',
 'CHIP:H3K4me2:A549 treated with 0.02% ethanol for 1 hour',
 'CHIP:H4K20me1:B cell female adult (27 years)',
 'CHIP:FOXP2:SK-N-MC',
 'CHIP:POLR2A:GM12878',
 'CHIP:SRF:GM12878',
 'CHIP:REST:GM12878',
 'CHIP:USF1:GM12878',
 'CHIP:JUND:HepG2',
 'CHIP:SIN3A:HepG2',
 'CHIP:USF1:HepG2',
 'CHIP:POLR2A:HeLa-S3',
 'CHIP:PBX3:GM12878',
 'CHIP:TAF1:GM12878',
 'CHIP:BATF:GM12878',
 'CHIP:EBF1:GM12878',
 'CHIP:IRF4:GM12878',
 'CHIP:TCF12:GM12878',
 'CHIP:BCL11A:GM12878',
 'CHIP:EP300:GM12878',
 'CHIP:ZBTB33:GM12878',
 'CHIP:PAX5:GM12878',
 'CHIP:NR3C1:A549 treated with 500 pM dexamethasone for 1 hour',
 'CHIP:NR3C1:A549 treated with 50 nM dexamethasone for 1 hour',
 'CHIP:NR3C1:A549 treated with 5 nM dexamethasone for 1 hour',
 'CHIP:POLR2A:A549 treated with 0.02% ethanol for 1 hour',
 'CHIP:POLR2A:A549 treated with 100 nM dexamethasone for 1 hour',
 'CHIP:PAX5:GM12878',
 'CHIP:REST:H1-hESC',
 'CHIP:POLR2A:H1-hESC',
 'CHIP:TAF1:H1-hESC',
 'CHIP:FOSL2:HepG2',
 'CHIP:GABPA:HeLa-S3',
 'CHIP:TAF1:HeLa-S3',
 'CHIP:RXRA:HepG2',
 'CHIP:POLR2A:GM12892',
 'CHIP:POLR2AphosphoS5:GM12892',
 'CHIP:TAF1:GM12892',
 'CHIP:POLR2AphosphoS5:H1-hESC',
 'CHIP:BHLHE40:HepG2',
 'CHIP:CTCF:HepG2',
 'CHIP:POLR2AphosphoS5:GM12878',
 'CHIP:POU2F2:GM12891',
 'CHIP:SPI1:GM12891',
 'CHIP:POLR2A:GM12891',
 'CHIP:POLR2AphosphoS5:GM12891',
 'CHIP:TAF1:GM12891',
 'CHIP:NR3C1:A549 treated with 0.02% ethanol for 1 hour',
 'CHIP:BCL11A:H1-hESC',
 'CHIP:SIX5:H1-hESC',
 'CHIP:SP1:H1-hESC',
 'CHIP:SIN3A:H1-hESC',
 'CHIP:TCF12:H1-hESC',
 'CHIP:USF1:H1-hESC',
 'CHIP:SRF:H1-hESC',
 'CHIP:GABPA:H1-hESC',
 'CHIP:ESR1:Ishikawa treated with 10 nM estradiol for 1 hour',
 'CHIP:ESR1:Ishikawa treated with 0.02% dimethyl sulfoxide for 1 hour',
 'CHIP:EGR1:H1-hESC',
 'CHIP:NR3C1:Ishikawa treated with 100 nM dexamethasone for 1 hour',
 'CHIP:RXRA:GM12878',
 'CHIP:SIX5:GM12878',
 'CHIP:TCF12:HepG2',
 'CHIP:PAX5:GM12891',
 'CHIP:REST:SK-N-SH',
 'CHIP:GABPA:HepG2',
 'CHIP:REST:HepG2',
 'CHIP:POLR2A:HepG2',
 'CHIP:TAF1:HepG2',
 'CHIP:REST:Panc1',
 'CHIP:REST:PFSK-1',
 'CHIP:NR3C1:A549 treated with 100 nM dexamethasone for 1 hour',
 'CHIP:ESR1:T47D treated with 100 nM genistein for 1 hour',
 'CHIP:NR3C1:A549 treated with 0.02% ethanol for 1 hour',
 'CHIP:RXRA:H1-hESC',
 'CHIP:SP1:HepG2',
 'CHIP:BCLAF1:GM12878',
 'CHIP:ETS1:GM12878',
 'CHIP:MEF2A:GM12878',
 'CHIP:ATF3:H1-hESC',
 'CHIP:YY1:H1-hESC',
 'CHIP:ATF3:HepG2',
 'CHIP:BCLAF1:K562',
 'CHIP:POLR2A:Ishikawa treated with 0.02% dimethyl sulfoxide for 1 hour',
 'CHIP:YY1:GM12891',
 'CHIP:EP300:H1-hESC',
 'CHIP:JUND:H1-hESC',
 'CHIP:ETS1:K562',
 'CHIP:POLR2AphosphoS5:K562',
 'CHIP:FOXA1:T47D treated with 0.02% dimethyl sulfoxide for 1 hour',
 'CHIP:NR3C1:Ishikawa treated with 0.02% ethanol for 1 hour',
 'CHIP:RAD21:H1-hESC',
 'CHIP:FOXA1:HepG2',
 'CHIP:HNF4A:HepG2',
 'CHIP:E2F6:K562',
 'CHIP:ESR1:T47D treated with 0.02% dimethyl sulfoxide for 1 hour',
 'CHIP:GABPA:K562',
 'CHIP:MAX:K562',
 'CHIP:SIN3A:K562',
 'CHIP:RAD21:HepG2',
 'CHIP:YY1:GM12892',
 'CHIP:TAF7:H1-hESC',
 'CHIP:EP300:HepG2',
 'CHIP:CTCF:SK-N-SH treated with 6 uM all-trans-retinoic acid for 48 hours',
 'CHIP:RAD21:SK-N-SH treated with 6 uM all-trans-retinoic acid for 48 hours',
 'CHIP:YY1:SK-N-SH treated with 6 uM all-trans-retinoic acid for 48 hours',
 'CHIP:HDAC2:HepG2',
 'CHIP:ZBTB7A:K562',
 'CHIP:USF1:SK-N-SH treated with 6 uM all-trans-retinoic acid for 48 hours',
 'CHIP:YY1:K562',
 'CHIP:SRF:GM12878',
 'CHIP:BCL11A:H1-hESC',
 'CHIP:POLR2AphosphoS5:HCT116',
 'CHIP:REST:HeLa-S3',
 'CHIP:FOXA1:HepG2',
 'CHIP:POLR2A:K562',
 'CHIP:NANOG:H1-hESC',
 'CHIP:FOSL1:K562',
 'CHIP:REST:K562',
 'CHIP:GATA3:T47D treated with 0.02% dimethyl sulfoxide for 1 hour',
 'CHIP:RAD21:GM12878',
 'CHIP:ELF1:HepG2',
 'CHIP:ZBTB33:HepG2',
 'CHIP:EGR1:K562',
 'CHIP:MEF2C:GM12878',
 'CHIP:CTCF:H1-hESC',
 'CHIP:FOXA2:HepG2',
 'CHIP:HNF4G:HepG2',
 'CHIP:TAF7:K562',
 'CHIP:THAP1:K562',
 'CHIP:YY1:GM12878',
 'CHIP:BCL3:GM12878',
 'CHIP:HDAC2:H1-hESC',
 'CHIP:FOSL1:H1-hESC',
 'CHIP:YY1:HepG2',
 'CHIP:ATF3:K562',
 ...]</code></pre>
</div>
</div>
<div class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>chrom_bed_downloads <span class="op">=</span> pd.read_csv(<span class="st">"https://uchicago.box.com/shared/static/du77wf31li38tciv8imivwu57svae03p.csv"</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>chrom_bed_downloads.index <span class="op">=</span> chrom_bed_downloads[<span class="st">"chroms"</span>]</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>chrom_bed_downloads.head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="15">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">chroms</th>
<th data-quarto-table-cell-role="th">link</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">chroms</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">1</td>
<td>1</td>
<td>https://uchicago.box.com/shared/static/9q9n4a0...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2</td>
<td>2</td>
<td>https://uchicago.box.com/shared/static/1tk6a3f...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">3</td>
<td>3</td>
<td>https://uchicago.box.com/shared/static/77ldwqq...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4</td>
<td>4</td>
<td>https://uchicago.box.com/shared/static/s0g48al...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">5</td>
<td>5</td>
<td>https://uchicago.box.com/shared/static/yafgxb1...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">6</td>
<td>6</td>
<td>https://uchicago.box.com/shared/static/9vpxc7z...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">7</td>
<td>7</td>
<td>https://uchicago.box.com/shared/static/hkru0gi...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">8</td>
<td>8</td>
<td>https://uchicago.box.com/shared/static/ruac33s...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9</td>
<td>9</td>
<td>https://uchicago.box.com/shared/static/dfw6gkj...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">10</td>
<td>10</td>
<td>https://uchicago.box.com/shared/static/ek50gvt...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>chr1_tss <span class="op">=</span> pd.read_table(<span class="st">'/grand/TFXcan/imlab/users/lvairus/hackenf/data/tss_by_chr/chr17_tss_by_gene.txt'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>amigo1_variations <span class="op">=</span> pd.read_table(<span class="st">'/grand/TFXcan/imlab/users/lvairus/hackenf/data/individual_beds/chr17/chr17_GSDMB.bed'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>geuvadis_gene_expression <span class="op">=</span> pd.read_table(<span class="st">'https://uchicago.box.com/shared/static/5vwc7pjw9qmtv7298c4rc7bcuicoyemt.gz'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>                                         dtype<span class="op">=</span>{<span class="st">'gene_id'</span>: <span class="bu">str</span>, <span class="st">'gene_name'</span>:<span class="bu">str</span>, <span class="st">'TargetID'</span>:<span class="bu">str</span>, <span class="st">'Chr'</span>:<span class="bu">str</span>})</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>geuvadis_gene_expression.head(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="16">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">gene_id</th>
<th data-quarto-table-cell-role="th">gene_name</th>
<th data-quarto-table-cell-role="th">TargetID</th>
<th data-quarto-table-cell-role="th">Chr</th>
<th data-quarto-table-cell-role="th">Coord</th>
<th data-quarto-table-cell-role="th">HG00096</th>
<th data-quarto-table-cell-role="th">HG00097</th>
<th data-quarto-table-cell-role="th">HG00099</th>
<th data-quarto-table-cell-role="th">HG00100</th>
<th data-quarto-table-cell-role="th">HG00101</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">NA20810</th>
<th data-quarto-table-cell-role="th">NA20811</th>
<th data-quarto-table-cell-role="th">NA20812</th>
<th data-quarto-table-cell-role="th">NA20813</th>
<th data-quarto-table-cell-role="th">NA20814</th>
<th data-quarto-table-cell-role="th">NA20815</th>
<th data-quarto-table-cell-role="th">NA20816</th>
<th data-quarto-table-cell-role="th">NA20819</th>
<th data-quarto-table-cell-role="th">NA20826</th>
<th data-quarto-table-cell-role="th">NA20828</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>ENSG00000223972.4</td>
<td>DDX11L1</td>
<td>ENSG00000223972.4</td>
<td>1</td>
<td>11869</td>
<td>0.320818</td>
<td>0.344202</td>
<td>0.354225</td>
<td>0.478064</td>
<td>-0.102815</td>
<td>...</td>
<td>1.008605</td>
<td>0.384489</td>
<td>0.581284</td>
<td>0.513981</td>
<td>0.667449</td>
<td>0.350890</td>
<td>0.186103</td>
<td>-0.037976</td>
<td>0.405439</td>
<td>0.199143</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>ENSG00000227232.3</td>
<td>WASH7P</td>
<td>ENSG00000227232.3</td>
<td>1</td>
<td>29806</td>
<td>33.714457</td>
<td>20.185174</td>
<td>18.095407</td>
<td>24.100871</td>
<td>29.018719</td>
<td>...</td>
<td>30.980194</td>
<td>34.086207</td>
<td>39.678442</td>
<td>29.643513</td>
<td>27.120420</td>
<td>29.121624</td>
<td>31.117198</td>
<td>32.047074</td>
<td>22.798959</td>
<td>23.563874</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>ENSG00000243485.1</td>
<td>MIR1302-11</td>
<td>ENSG00000243485.1</td>
<td>1</td>
<td>29554</td>
<td>0.240408</td>
<td>0.157456</td>
<td>0.218806</td>
<td>0.320878</td>
<td>0.067833</td>
<td>...</td>
<td>0.065940</td>
<td>0.228784</td>
<td>0.140642</td>
<td>0.283905</td>
<td>0.273821</td>
<td>0.286311</td>
<td>0.324060</td>
<td>0.049574</td>
<td>0.255288</td>
<td>0.157440</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>ENSG00000238009.2</td>
<td>RP11-34P13.7</td>
<td>ENSG00000238009.2</td>
<td>1</td>
<td>133566</td>
<td>0.328272</td>
<td>0.327932</td>
<td>0.090064</td>
<td>0.420443</td>
<td>0.220269</td>
<td>...</td>
<td>0.274071</td>
<td>0.384179</td>
<td>0.533693</td>
<td>0.307221</td>
<td>0.307367</td>
<td>0.400278</td>
<td>0.612321</td>
<td>0.666633</td>
<td>0.281138</td>
<td>1.346129</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>ENSG00000239945.1</td>
<td>RP11-34P13.8</td>
<td>ENSG00000239945.1</td>
<td>1</td>
<td>91105</td>
<td>0.332171</td>
<td>-0.032164</td>
<td>0.017323</td>
<td>0.424677</td>
<td>0.214025</td>
<td>...</td>
<td>0.347323</td>
<td>0.346744</td>
<td>0.073580</td>
<td>0.400396</td>
<td>0.470517</td>
<td>0.069749</td>
<td>0.299353</td>
<td>0.090019</td>
<td>0.282554</td>
<td>-0.157170</td>
</tr>
</tbody>
</table>

<p>5 rows × 467 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> Enformer(model_path) <span class="co"># here we load the model architecture.</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>fasta_extractor <span class="op">=</span> FastaStringExtractor(fasta_file) <span class="co"># we define a class called fasta_extractor to help us extra raw sequence data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>gene_intervals <span class="op">=</span> collect_intervals(chromosomes<span class="op">=</span>[<span class="st">'17'</span>], gene_list<span class="op">=</span>[<span class="st">'GSDMB'</span>])</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gene_intervals)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'GSDMB': ['17', 38060848, 38077313]}</code></pre>
</div>
</div>
</section>
<section id="run-predictions" class="level2">
<h2 class="anchored" data-anchor-id="run-predictions">Run Predictions</h2>
<p>We’ll pick one individual at random.</p>
<div class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rand_individual = np.random.choice(a=geuvadis_gene_expression.columns[6:-1], replace=False) # individuals we are interested in</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>rand_individual <span class="op">=</span> <span class="st">'NA12413'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>test <span class="op">=</span> My_class(<span class="st">'GSDMB'</span>, <span class="st">'17'</span>, rand_individual)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>diffmats <span class="op">=</span> test.get_diffmats()</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="bu">abs</span>, rel, relmax, relmed <span class="op">=</span> diffmats</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>rel</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="21">
<pre><code>array([[2.68207751e-02, 2.14794427e-02, 1.54860010e-02, ...,
        3.27786431e-03, 4.31587268e-03, 5.99253085e-03],
       [1.52227283e-02, 9.49528441e-03, 1.39587221e-03, ...,
        4.25775489e-03, 6.33378932e-03, 3.99001292e-04],
       [5.20520937e-03, 5.04603982e-03, 1.07796649e-02, ...,
        2.46862252e-03, 5.82935056e-03, 1.01276860e-03],
       ...,
       [1.82648703e-01, 1.20395273e-01, 2.26677638e-02, ...,
        3.71882971e-03, 2.21709910e-04, 3.99682624e-03],
       [7.65870988e-01, 4.11944836e-01, 7.03058019e-03, ...,
        7.04148132e-03, 2.09397208e-02, 2.70997807e-02],
       [4.41073626e-01, 1.37864158e-01, 3.22264820e-01, ...,
        1.49025135e-02, 2.87779793e-03, 4.64532932e-04]], dtype=float32)</code></pre>
</div>
</div>
<div class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>test.get_summary(test.rel_diffmat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="22">
<pre><code>{'mean': 0.21557848,
 'median': 0.08500237,
 'std_dev': 0.5238661,
 'minimum': 0.0,
 'maximum': 73.63812,
 'total_sum': 1026250.1,
 'q1': 0.02724431036040187,
 'q2': 0.0850023701786995,
 'q3': 0.21821939945220947,
 'iqr': 0.1909750890918076}</code></pre>
</div>
</div>
<div class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>abs_table <span class="op">=</span> test.make_table(<span class="bu">abs</span>, <span class="dv">1</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>abs_table_sorted <span class="op">=</span> abs_table.sort_values(<span class="st">'diff'</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>abs_table_sorted</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="23">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">track</th>
<th data-quarto-table-cell-role="th">bin</th>
<th data-quarto-table-cell-role="th">diff</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">789</td>
<td>CHIP:H3K4me1:thymus male child (3 years)</td>
<td>533</td>
<td>1.000574</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2468</td>
<td>CHIP:H3K4me3:esophagus male adult (34 years)</td>
<td>760</td>
<td>1.000663</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2393</td>
<td>CHIP:HNRNPLL:K562</td>
<td>760</td>
<td>1.000681</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2649</td>
<td>CHIP:H3K4me1:CD14-positive monocyte male adult...</td>
<td>867</td>
<td>1.000864</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">220</td>
<td>CHIP:CTCF:A549 treated with 100 nM dexamethaso...</td>
<td>134</td>
<td>1.001465</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1002</td>
<td>CHIP:CTCF:K562</td>
<td>540</td>
<td>7.443279</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1070</td>
<td>CHIP:CTCF:K562</td>
<td>540</td>
<td>7.470882</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">864</td>
<td>CHIP:CTCF:K562</td>
<td>539</td>
<td>8.511012</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1674</td>
<td>CAGE:thymus, adult, pool1</td>
<td>563</td>
<td>10.100098</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1690</td>
<td>CAGE:thymus, fetal, pool1</td>
<td>563</td>
<td>10.273163</td>
</tr>
</tbody>
</table>

<p>2669 rows × 3 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> mat <span class="kw">in</span> diffmats:</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(test.get_summary(mat))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>{'mean': 0.011740582, 'median': 0.002480626, 'std_dev': 0.065571606, 'minimum': 0.0, 'maximum': 10.273163, 'total_sum': 55890.43, 'q1': 0.0005608946084976196, 'q2': 0.002480626106262207, 'q3': 0.007642507553100586, 'iqr': 0.007081612944602966}
{'mean': 0.21557848, 'median': 0.08500237, 'std_dev': 0.5238661, 'minimum': 0.0, 'maximum': 73.63812, 'total_sum': 1026250.1, 'q1': 0.02724431036040187, 'q2': 0.0850023701786995, 'q3': 0.21821939945220947, 'iqr': 0.1909750890918076}
{'mean': 0.03527108, 'median': 0.0074218167, 'std_dev': 0.08494974, 'minimum': 0.0, 'maximum': 1.0, 'total_sum': 167906.14, 'q1': 0.0011184182367287576, 'q2': 0.007421816699206829, 'q3': 0.03162767644971609, 'iqr': 0.030509258212987334}
{'mean': 5.7709837, 'median': 1.0, 'std_dev': 116.90829, 'minimum': 0.0, 'maximum': 36243.305, 'total_sum': 27472468.0, 'q1': 0.38562167435884476, 'q2': 1.0, 'q3': 2.418138027191162, 'iqr': 2.0325163528323174}</code></pre>
</div>
</div>
<div class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(test.get_outliers(rel, <span class="dv">8</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">
<pre><code>2285</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>values <span class="op">=</span> np.linspace(<span class="dv">1</span>, <span class="dv">80</span>, <span class="dv">10</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> v <span class="kw">in</span> values:</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="bu">len</span>(test.get_outliers()))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>test.plot_tol_numouts(rel, <span class="dv">1</span>, <span class="dv">80</span>, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Compare_averages_files/figure-html/cell-28-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>test_abs_out <span class="op">=</span> test.get_outliers(test.abs_mat, <span class="dv">1</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>test.get_outliers(test.abs_mat, <span class="dv">1</span>), <span class="bu">len</span>(test.get_outliers(test.abs_mat, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="21">
<pre><code>(array([1.0394936, 1.0977535, 1.0580368, ..., 1.274787 , 1.6978073,
        1.5110455], dtype=float32),
 2669)</code></pre>
</div>
</div>
<div class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>test.make_hist(test_abs_out, <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Compare_averages_files/figure-html/cell-30-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>test.get_outlier_inds(test.abs_mat, <span class="dv">1</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>col_outs <span class="op">=</span> test.get_outlier_inds(test.abs_mat, <span class="dv">1</span>)[<span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>to get list of unique inds: set(inds_arr) 
get count of inds: Counter(inds_arr)
to get list of unique inds: set(inds_arr) 
get count of inds: Counter(inds_arr)</code></pre>
</div>
</div>
<div class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>test.make_hist(col_outs, <span class="dv">500</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Compare_averages_files/figure-html/cell-32-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>outlier_df <span class="op">=</span> test.make_table(test.abs_mat, <span class="dv">1</span>)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>outlier_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="29">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">track</th>
<th data-quarto-table-cell-role="th">bin</th>
<th data-quarto-table-cell-role="th">diff</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>CHIP:H3K27ac:OCI-LY3</td>
<td>44</td>
<td>1.039494</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>CHIP:H3K27ac:OCI-LY3</td>
<td>45</td>
<td>1.097754</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>CHIP:H3K27ac:OCI-LY3</td>
<td>46</td>
<td>1.058037</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>CHIP:H3K27ac:OCI-LY3</td>
<td>48</td>
<td>1.166883</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>CHIP:H3K27ac:OCI-LY3</td>
<td>53</td>
<td>1.050079</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2664</td>
<td>CHIP:H3K4me1:CD14-positive monocyte male adult...</td>
<td>893</td>
<td>1.130640</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2665</td>
<td>CHIP:H3K4me1:CD14-positive monocyte female</td>
<td>894</td>
<td>1.405522</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2666</td>
<td>CHIP:H3K4me1:CD14-positive monocyte male adult...</td>
<td>894</td>
<td>1.274787</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2667</td>
<td>CHIP:H3K4me1:CD14-positive monocyte female</td>
<td>895</td>
<td>1.697807</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2668</td>
<td>CHIP:H3K4me1:CD14-positive monocyte male adult...</td>
<td>895</td>
<td>1.511045</td>
</tr>
</tbody>
</table>

<p>2669 rows × 3 columns</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>    <span class="co"># def get_summary(self, arr):</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># def get_outliers(self, arr, tol):</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>    <span class="co"># def make_hist(arr, bin_num):</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># def get_outlier_inds(self, arr, tol):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>test2 <span class="op">=</span> My_class(<span class="st">"AMIGO1"</span>, <span class="st">"1"</span>, rand_individual)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>test2_abs_diff <span class="op">=</span> test2.get_abs_diff()</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>test2_abs_diff, <span class="bu">type</span>(test2_abs_diff)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="42">
<pre><code>(array([[5.7518482e-06, 3.5203993e-06, 8.2701445e-07, ..., 6.4517371e-07,
         6.3516200e-07, 1.2149103e-06],
        [2.5629997e-06, 1.7061830e-06, 6.2957406e-06, ..., 3.6321580e-08,
         3.4179538e-06, 1.1073425e-06],
        [1.4603138e-06, 1.5571713e-06, 1.6354024e-06, ..., 1.9324943e-08,
         2.0535663e-06, 6.4587221e-07],
        ...,
        [5.5428594e-05, 6.6407025e-05, 5.4815784e-05, ..., 1.5444122e-05,
         2.6834011e-04, 1.5201420e-04],
        [9.3728304e-05, 1.3028085e-04, 1.2800470e-04, ..., 4.4204295e-05,
         4.0826201e-04, 2.2485107e-04],
        [3.8206577e-05, 5.6162477e-05, 9.3773007e-05, ..., 4.2296946e-05,
         5.4708123e-04, 2.2122264e-04]], dtype=float32),
 numpy.ndarray)</code></pre>
</div>
</div>
<div class="cell" data-execution_count="44">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>abs_diff2_sum <span class="op">=</span> test2.get_summary(test2_abs_diff)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>abs_diff2_sum</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="44">
<pre><code>{'mean': 0.0008395509,
 'median': 0.00011945516,
 'std_dev': 0.009009555,
 'minimum': 0.0,
 'maximum': 2.1459656,
 'total_sum': 3996.6384,
 'q1': 2.8625130653381348e-05,
 'q2': 0.00011945515871047974,
 'q3': 0.0003998279571533203,
 'iqr': 0.00037120282649993896}</code></pre>
</div>
</div>
<div class="cell" data-execution_count="45">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>test2.get_summary(test2.abs_mat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="45">
<pre><code>{'mean': 0.0008395509,
 'median': 0.00011945516,
 'std_dev': 0.009009555,
 'minimum': 0.0,
 'maximum': 2.1459656,
 'total_sum': 3996.6384,
 'q1': 2.8625130653381348e-05,
 'q2': 0.00011945515871047974,
 'q3': 0.0003998279571533203,
 'iqr': 0.00037120282649993896}</code></pre>
</div>
</div>
<div class="cell" data-execution_count="46">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>abs_outs2 <span class="op">=</span> test2.get_outliers(test2.abs_mat, <span class="dv">1</span>)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(abs_outs2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="46">
<pre><code>78</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>test2.get_outlier_inds(test2.abs_mat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>test2.make_hist(abs_outs2, <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Compare_averages_files/figure-html/cell-41-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="107">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>gene <span class="op">=</span> <span class="st">'GSDMB'</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>gene_interval <span class="op">=</span> gene_intervals[gene]</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>target_interval <span class="op">=</span> kipoiseq.Interval(<span class="st">"chr"</span> <span class="op">+</span> gene_interval[<span class="dv">0</span>],</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>                                        gene_interval[<span class="dv">1</span>],</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>                                        gene_interval[<span class="dv">2</span>])</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>target_fa <span class="op">=</span> fasta_extractor.extract(target_interval.resize(SEQUENCE_LENGTH))</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>window_coords <span class="op">=</span> target_interval.resize(SEQUENCE_LENGTH)</span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>cur_gene_vars <span class="op">=</span> pd.read_csv(<span class="st">"/grand/TFXcan/imlab/users/lvairus/hackenf/data/individual_beds/chr"</span> <span class="op">+</span> gene_interval[<span class="dv">0</span>] <span class="op">+</span> <span class="st">"/chr"</span> <span class="op">+</span> gene_interval[<span class="dv">0</span>] <span class="op">+</span> <span class="st">"_"</span><span class="op">+</span> gene <span class="op">+</span> <span class="st">".bed"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, header<span class="op">=</span><span class="dv">0</span>) <span class="co"># read in the appropriate bed file for the gene</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="108">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>haplo_1, haplo_2 <span class="op">=</span> geno_to_seq(gene, rand_individual)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>haplo_1_enc <span class="op">=</span> one_hot_encode(<span class="st">""</span>.join(haplo_1))[np.newaxis]</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>haplo_2_enc <span class="op">=</span> one_hot_encode(<span class="st">""</span>.join(haplo_2))[np.newaxis]</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>average_enc <span class="op">=</span> np.add(haplo_1_enc, haplo_2_enc) <span class="op">/</span> <span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="109">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>prediction_1 <span class="op">=</span> model.predict_on_batch(haplo_1_enc)[<span class="st">'human'</span>][<span class="dv">0</span>]</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>prediction_2 <span class="op">=</span> model.predict_on_batch(haplo_2_enc)[<span class="st">'human'</span>][<span class="dv">0</span>]</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>post_average <span class="op">=</span> (prediction_1 <span class="op">+</span> prediction_2) <span class="op">/</span> <span class="dv">2</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>pre_average <span class="op">=</span> model.predict_on_batch(average_enc)[<span class="st">'human'</span>][<span class="dv">0</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="comparing-predictions" class="level2">
<h2 class="anchored" data-anchor-id="comparing-predictions">Comparing Predictions</h2>
<div class="cell" data-execution_count="113">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>pre_average2, post_average2 <span class="op">=</span> run_predictions2(gene, <span class="st">'17'</span>, rand_individual)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Currently on gene GSDMB, and predicting on individual NA12413...</code></pre>
</div>
</div>
<div class="cell" data-execution_count="110">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>diff <span class="op">=</span> pre_average <span class="op">-</span> post_average</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>abs_diff <span class="op">=</span> np.sqrt(np.square(diff))</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>rel_diff <span class="op">=</span> (abs_diff) <span class="op">/</span> ((pre_average <span class="op">+</span> post_average) <span class="op">+</span> <span class="dv">10</span><span class="op">**-</span><span class="dv">16</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="111">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Summary Statistics of Absolute Differece</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>arr <span class="op">=</span> abs_diff</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Mean:"</span>, np.mean(arr))</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Median:"</span>, np.median(arr))</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Standard Deviation:"</span>, np.std(arr))</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Minimum:"</span>, np.<span class="bu">min</span>(arr))</span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Maximum:"</span>, np.<span class="bu">max</span>(arr))</span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Sum:"</span>, np.<span class="bu">sum</span>(arr))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Mean: 0.011741725
Median: 0.0024814606
Standard Deviation: 0.06555718
Minimum: 0.0
Maximum: 10.2612915
Sum: 55895.87</code></pre>
</div>
</div>
<div class="cell" data-execution_count="91">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Summary Statistics of Relative Difference</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>arr <span class="op">=</span> rel_diff</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Mean:"</span>, np.mean(arr))</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Median:"</span>, np.median(arr))</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Standard Deviation:"</span>, np.std(arr))</span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Minimum:"</span>, np.<span class="bu">min</span>(arr))</span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Maximum:"</span>, np.<span class="bu">max</span>(arr))</span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Sum:"</span>, np.<span class="bu">sum</span>(arr))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Mean: 0.00026652514
Median: 8.4674466e-05
Standard Deviation: 0.0011749842
Minimum: 0.0
Maximum: 0.24099354
Sum: 1268.779</code></pre>
</div>
</div>
<div class="cell" data-execution_count="92">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>bigDiff <span class="op">=</span> abs_diff[abs_diff<span class="op">&gt;</span><span class="dv">1</span>]</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(bigDiff)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="92">
<pre><code>72</code></pre>
</div>
</div>
<div class="cell" data-execution_count="93">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the histogram</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>plt.hist(bigDiff, bins<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Add labels and title</span></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">'Value'</span>)</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'Frequency'</span>)</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'Histogram'</span>)</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the plot</span></span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Compare_averages_files/figure-html/cell-50-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="in-which-tracks-is-it-not-precise" class="level2">
<h2 class="anchored" data-anchor-id="in-which-tracks-is-it-not-precise">In which tracks is it not precise?</h2>
<div class="cell" data-execution_count="94">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># max diff tolerance is 1. If a diff is greater than 1 we will count it as too big</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>tolerance <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>indices <span class="op">=</span> np.where(abs_diff <span class="op">&gt;</span> tolerance)</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>indices[<span class="dv">0</span>][<span class="dv">0</span>], indices[<span class="dv">1</span>][<span class="dv">0</span>]</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>ind_of_big_diffs <span class="op">=</span> np.array(<span class="bu">list</span>(<span class="bu">zip</span>(indices[<span class="dv">0</span>], indices[<span class="dv">1</span>])))</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(ind_of_big_diffs), ind_of_big_diffs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="94">
<pre><code>(72,
 array([[ 773, 4740],
        [ 773, 4746],
        [ 773, 4747],
        [ 773, 4748],
        [ 773, 4754],
        [ 773, 4759],
        [ 773, 4760],
        [ 773, 4764],
        [ 773, 4766],
        [ 773, 4770],
        [ 773, 4777],
        [ 773, 4782],
        [ 773, 4797],
        [ 773, 4808],
        [ 773, 4810],
        [ 773, 4815],
        [ 773, 4816],
        [ 773, 4817],
        [ 773, 4818],
        [ 773, 4819],
        [ 773, 4868],
        [ 773, 4869],
        [ 773, 4870],
        [ 773, 4874],
        [ 773, 4877],
        [ 773, 4879],
        [ 773, 4881],
        [ 773, 4884],
        [ 773, 4885],
        [ 773, 4886],
        [ 773, 4887],
        [ 773, 4890],
        [ 773, 4892],
        [ 773, 4893],
        [ 773, 4897],
        [ 773, 4898],
        [ 773, 4900],
        [ 773, 4902],
        [ 773, 4905],
        [ 773, 4906],
        [ 773, 4909],
        [ 773, 4920],
        [ 773, 5056],
        [ 773, 5057],
        [ 773, 5072],
        [ 773, 5073],
        [ 773, 5078],
        [ 773, 5079],
        [ 773, 5080],
        [ 773, 5106],
        [ 773, 5110],
        [ 773, 5115],
        [ 773, 5127],
        [ 773, 5144],
        [ 773, 5150],
        [ 773, 5196],
        [ 773, 5198],
        [ 773, 5212],
        [ 773, 5213],
        [ 773, 5236],
        [ 773, 5238],
        [ 773, 5300],
        [ 829, 2652],
        [ 830,  733],
        [ 830,  746],
        [ 830,  801],
        [ 830,  810],
        [ 830,  812],
        [ 830,  814],
        [ 830,  815],
        [ 830,  822],
        [ 830, 3647]]))</code></pre>
</div>
</div>
<section id="cols" class="level3">
<h3 class="anchored" data-anchor-id="cols">cols</h3>
<div class="cell" data-execution_count="95">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># all the indices of the columns where the diff exceeds tolerance</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>col_inds <span class="op">=</span> indices[<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="96">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>counts <span class="op">=</span> Counter(col_inds)</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>[(key,value) <span class="cf">for</span> key,value <span class="kw">in</span> counts.items() <span class="cf">if</span> value <span class="op">&gt;</span> <span class="dv">8</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="96">
<pre><code>[]</code></pre>
</div>
</div>
<div class="cell" data-execution_count="97">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Number of unique col ind: </span><span class="sc">{</span><span class="bu">len</span>(<span class="bu">set</span>(col_inds))<span class="sc">}</span><span class="ss"> </span><span class="ch">\n</span><span class="ss">Total col ind: </span><span class="sc">{</span><span class="bu">len</span>(col_inds)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of unique col ind: 72 
Total col ind: 72</code></pre>
</div>
</div>
<div class="cell" data-execution_count="98">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>plt.hist(col_inds, bins<span class="op">=</span><span class="dv">200</span>)  <span class="co"># Specify the number of bins</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'Histogram of col indexes'</span>)</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Compare_averages_files/figure-html/cell-55-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>Analysis: the difference exceeds the tolerance in 227 unique columns that are pretty evenly distributed, so enformer doesn’t necessarily do better or worse with averaging before/after in any particular cell line</p>
<p>i want to get the amount of times it affects each col 227 uniqe cols</p>
</section>
<section id="rows" class="level3">
<h3 class="anchored" data-anchor-id="rows">rows</h3>
<div class="cell" data-execution_count="99">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(<span class="bu">set</span>(indices[<span class="dv">0</span>])), <span class="bu">len</span>(indices[<span class="dv">0</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="99">
<pre><code>(3, 72)</code></pre>
</div>
</div>
<div class="cell" data-execution_count="100">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Summary Statistics of rows</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>arr1 <span class="op">=</span> indices[<span class="dv">0</span>]</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Mean:"</span>, np.mean(arr1))</span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Median:"</span>, np.median(arr1))</span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Standard Deviation:"</span>, np.std(arr1))</span>
<span id="cb78-7"><a href="#cb78-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Minimum:"</span>, np.<span class="bu">min</span>(arr1))</span>
<span id="cb78-8"><a href="#cb78-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Maximum:"</span>, np.<span class="bu">max</span>(arr1))</span>
<span id="cb78-9"><a href="#cb78-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Sum:"</span>, np.<span class="bu">sum</span>(arr1))</span>
<span id="cb78-10"><a href="#cb78-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-11"><a href="#cb78-11" aria-hidden="true" tabindex="-1"></a>q1 <span class="op">=</span> np.percentile(arr, <span class="dv">25</span>)</span>
<span id="cb78-12"><a href="#cb78-12" aria-hidden="true" tabindex="-1"></a>q3 <span class="op">=</span> np.percentile(arr, <span class="dv">75</span>)</span>
<span id="cb78-13"><a href="#cb78-13" aria-hidden="true" tabindex="-1"></a>iqr <span class="op">=</span> q3 <span class="op">-</span> q1</span>
<span id="cb78-14"><a href="#cb78-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-15"><a href="#cb78-15" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Q1:"</span>, q1)</span>
<span id="cb78-16"><a href="#cb78-16" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Q3:"</span>, q3)</span>
<span id="cb78-17"><a href="#cb78-17" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Interquartile Range:"</span>, iqr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Mean: 780.9027777777778
Median: 773.0
Standard Deviation: 19.678075590631757
Minimum: 773
Maximum: 830
Sum: 56225</code></pre>
</div>
</div>
<div class="cell" data-execution_count="101">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>plt.hist(indices[<span class="dv">0</span>], bins<span class="op">=</span><span class="dv">30</span>)  <span class="co"># Specify the number of bins</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'Histogram of row indexes'</span>)</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="Compare_averages_files/figure-html/cell-58-output-1.png" class="img-fluid"></p>
</div>
</div>
<p>Analysis: There are only a few specific locations where enformer does worse</p>
</section>
</section>
<section id="comparing-across-tracks" class="level2">
<h2 class="anchored" data-anchor-id="comparing-across-tracks">Comparing across tracks</h2>
<div class="cell" data-execution_count="102">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>res <span class="op">=</span> []</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">5313</span>):</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>    pre_track <span class="op">=</span> pre_average[:, i]</span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>    post_track <span class="op">=</span> post_average[:, i]</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>    corr <span class="op">=</span> np.corrcoef(pre_track, post_track)[<span class="dv">0</span>][<span class="dv">1</span>]</span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a>    res.append(corr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The results from both methods are nearly identical.</p>
<div class="cell" data-execution_count="103">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">min</span>(res), <span class="bu">max</span>(res))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0.9989958894392978 0.9999999912865261</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>