<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Laura Vairus">
<meta name="dcterms.date" content="2023-06-06">
<meta name="description" content="Summary of LLM article by Serafim Batzoglou">

<title>Laura’s Blog - Large Language Models</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Laura’s Blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com" rel="" target=""><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Large Language Models</h1>
                  <div>
        <div class="description">
          Summary of LLM article by Serafim Batzoglou
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Laura Vairus </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">June 6, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<p><img src="https://miro.medium.com/v2/resize:fit:1400/format:webp/1*MZus-glUYW6V1ucTH2A1EQ.png" class="img-fluid"></p>
<p>In his article “Large Language Models in Molecular Biology,” Serafim Batzoglou discusses recent advancements in deep learning-based language models and their potential to significantly impact the field of molecular biology. The convergence of LLMs with large-scale genomic and population health data is expected to propel the understanding and modeling of biomolecular systems with a level of accuracy that surpasses human capacity.</p>
<section id="large-language-models-llms" class="level3">
<h3 class="anchored" data-anchor-id="large-language-models-llms">Large Language Models (LLMs):</h3>
<p>LLMs are advanced neural networks capable of generating text that resembles human language. They’re trained with vast amounts of data and learn to predict subsequent words in a sentence based on preceding words, allowing them to understand patterns, relationships, and context within the text.</p>
</section>
<section id="the-genetic-dogma" class="level3">
<h3 class="anchored" data-anchor-id="the-genetic-dogma">The Genetic Dogma:</h3>
<p><img src="https://miro.medium.com/v2/resize:fit:1100/format:webp/1*xh5qab9iXVx21oblKKbc2w.jpeg" class="img-fluid" width="702"></p>
<p>The central dogma of molecular biology describes the unidirectional flow of genetic information from DNA to RNA to protein. Within the DNA, there are approximately 20,000 genes responsible for synthesizing proteins, which play crucial roles in various biological processes: acting as structural components, enzymes, and facilitating communication within cells.</p>
<p>The process of protein synthesis includes transcription, splicing, and translation. Transcription produces messenger RNA (mRNA) by copying a DNA segment, which is then spliced to remove introns and retain exons, forming mature mRNA. Translation decodes the mRNA sequence into amino acids, which are linked to form proteins.</p>
<p>Gene regulation, mediated by transcription factors and other proteins, ensures the timely and appropriate expression of genes within cells. Chromatin structure, consisting of DNA wrapped around histone proteins, plays a role in gene accessibility and regulation; and histone modifications and DNA methylation influence gene expression by changing chromatin structure.</p>
</section>
<section id="genetic-variation" class="level3">
<h3 class="anchored" data-anchor-id="genetic-variation">Genetic Variation:</h3>
<p>An individuals’ DNA and environmental influences shapes their biology throughout their lifetime. While humans share over 99.9% identical DNA, our DNA variants account for the heritability of traits, including contributions to health and disease.</p>
<p>DNA variants are introduced through mutations in the DNA inherited from parents. Most variants are benign, while some may be deleterious or beneficial. Deleterious variants tend to be statistically eliminated from the population over time through natural selection. Essential genes are conserved with little to no mutation across generations due to their crucial role in fitness. Conversely, less significant genes are prone to more mutations as they have minimal fitness consequences and can be passed down without significant impact.</p>
<p>Advancements in DNA sequencing technologies have made data collection rapid and cost-effective. Sequencing-based methods can measure various molecular functions, such as gene expression and chromatin structure. Other technologies like mass spectrometry and X-ray crystallography provide insights into protein levels and structures.</p>
<p>Genome-wide association studies (GWAS) correlate genetic variants with specific phenotypes, providing valuable insights into gene functions and disease mechanisms. However, large language models (LLMs) are expected to surpass traditional association analyses in linking genetic variation to function. The combination of DNA sequencing technologies, data generation capabilities, and LLMs holds promise in advancing our understanding of genetic variation and its impact on molecular mechanisms and human physiology.</p>
</section>
<section id="prominent-language-models-in-molecular-biology" class="level3">
<h3 class="anchored" data-anchor-id="prominent-language-models-in-molecular-biology">Prominent Language Models in Molecular Biology:</h3>
<p>In recent years, significant progress has been made in modeling the central dogma of molecular biology, offering insights into gene function and expression. While fully transforming molecular biology into a computational science or engineering human health is still a work in progress, the current momentum suggests that it’s achievable with more data and development. For example, language models like LLMs excel at learning intricate statistical properties of complex sequential data.</p>
<p>Breakthroughs in different stages of the central dogma exemplify this progress. Methods like SpliceAI accurately predict gene structure by identifying splicing sites, aiding in genetic disease diagnosis. Advances in protein structure prediction, particularly with AlphaFold, have come close to solving the protein folding problem, revolutionizing biological research and drug discovery. Tools like PrimateAI-3D help annotate genetic variants as benign or pathogenic, contributing to disease diagnosis and drug target identification. Language models such as Enformer show promise in predicting gene expression from DNA sequences alone, shedding light on gene regulation. Foundation models like scGPT and Nucleotide Transformer, trained on extensive data, provide valuable insights into single-cell biology and raw DNA sequences, facilitating various downstream applications.</p>
<p><img src="https://miro.medium.com/v2/resize:fit:1100/format:webp/1*BRVmjtCysN8Sow3_X4n-zQ.jpeg" class="img-fluid" width="701"></p>
<p>Overall, the progress in modeling the central dogma of molecular biology demonstrates the potential of AI and deep learning in understanding gene function, expression, and regulation. Continued advancements and integration of diverse data sources will enhance our understanding of molecular biology and its impact on medicine and human health.</p>
</section>
<section id="looking-forward" class="level3">
<h3 class="anchored" data-anchor-id="looking-forward">Looking Forward</h3>
<p>The availability of rich and affordable data is a significant driver of progress in this field. Advances in DNA sequencing technology have significantly reduced the cost of genome sequencing and other molecular assays, enabling comprehensive profiling of gene expression, chromatin structure, and other molecular layers. Initiatives like the UK Biobank and All Of Us project have collected extensive genetic and health data from large cohorts of participants, offering a wealth of information for research purposes. Cancer-focused companies, such as Tempus and Foundation Medicine, are also building vast genomic databases with clinical information.</p>
<p>While developing these technologies, privacy concerns and the role of LLMs in clinical practice must be addressed. Proper informed consent and privacy measures are essential when training LLMs with participant data. LLMs should be seen as tools to assist healthcare professionals rather than replace them. Patient trust and verification remain crucial aspects of their application in clinical settings.</p>
<p>LLMs are well-suited to integrate and analyze these diverse datasets. Through autoregressive or masked language modeling, these models can learn from single-cell datasets, functional genomics data, and clinical records to understand gene pathways, genomic variants, and their associations with human health. However, the challenge lies in technical innovations to represent and integrate different layers of information and scale up the model’s processing capacity.</p>
<p>Such an advanced LLM could have various applications, including clinical diagnosis, drug development, and advancing our understanding of molecular biology. It could aid doctors in making precise diagnoses, identify potential drug targets, and assist in personalized medicine. Additionally, the model could offer suggestions for additional experiments and contribute to filling gaps in data.</p>
</section>
<section id="conclusion" class="level3">
<h3 class="anchored" data-anchor-id="conclusion">Conclusion</h3>
<p>Molecular biology’s transition into a computational science is facilitated by the combination of extensive data acquisition and powerful models like LLMs. In the coming years, these models will enable accurate predictions of the complex biomolecular interactions that connect our DNA, cellular biology, and overall health. This advancement is expected to have a profound impact on various aspects of medicine. Additionally, the development of open foundation models that integrate genomic and medical data will drive research, innovation, and facilitate the practice of precision medicine.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>